{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 0,
   "metadata": {
    "application/vnd.databricks.v1+cell": {
     "cellMetadata": {
      "byteLimit": 2048000,
      "rowLimit": 10000
     },
     "inputWidgets": {},
     "nuid": "9f97a76c-946b-4ce9-b476-ac9c7e2b500f",
     "showTitle": false,
     "title": ""
    }
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001B[43mNote: you may need to restart the kernel using dbutils.library.restartPython() to use updated packages.\u001B[0m\nRequirement already satisfied: tqdm in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (4.66.4)\nRequirement already satisfied: dask in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (2024.5.2)\nRequirement already satisfied: cairosvg in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (2.7.1)\nRequirement already satisfied: pillow in /databricks/python3/lib/python3.10/site-packages (9.2.0)\nRequirement already satisfied: matplotlib in /databricks/python3/lib/python3.10/site-packages (3.5.2)\nRequirement already satisfied: easyocr in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (1.7.1)\nRequirement already satisfied: scikit-learn in /databricks/python3/lib/python3.10/site-packages (1.1.1)\nRequirement already satisfied: opencv-python-headless in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (4.10.0.82)\nRequirement already satisfied: pandas in /databricks/python3/lib/python3.10/site-packages (1.4.4)\nRequirement already satisfied: distributed in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (2024.5.2)\nRequirement already satisfied: torch in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (2.3.1)\nRequirement already satisfied: torchvision in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (0.18.1)\nRequirement already satisfied: torchaudio in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (2.3.1)\nRequirement already satisfied: pyspark in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (3.5.1)\nRequirement already satisfied: setuptools in /databricks/python3/lib/python3.10/site-packages (63.4.1)\nRequirement already satisfied: boto3 in /databricks/python3/lib/python3.10/site-packages (1.24.28)\nRequirement already satisfied: s3fs in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (2023.3.0)\nRequirement already satisfied: fsspec>=2021.09.0 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from dask) (2023.3.0)\nRequirement already satisfied: partd>=1.2.0 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from dask) (1.4.2)\nRequirement already satisfied: cloudpickle>=1.5.0 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from dask) (3.0.0)\nRequirement already satisfied: click>=8.1 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from dask) (8.1.7)\nRequirement already satisfied: pyyaml>=5.3.1 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from dask) (6.0.1)\nRequirement already satisfied: importlib-metadata>=4.13.0 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from dask) (7.1.0)\nRequirement already satisfied: toolz>=0.10.0 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from dask) (0.12.1)\nRequirement already satisfied: packaging>=20.0 in /databricks/python3/lib/python3.10/site-packages (from dask) (21.3)\nRequirement already satisfied: defusedxml in /databricks/python3/lib/python3.10/site-packages (from cairosvg) (0.7.1)\nRequirement already satisfied: tinycss2 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from cairosvg) (1.3.0)\nRequirement already satisfied: cssselect2 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from cairosvg) (0.7.0)\nRequirement already satisfied: cairocffi in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from cairosvg) (1.7.0)\nRequirement already satisfied: fonttools>=4.22.0 in /databricks/python3/lib/python3.10/site-packages (from matplotlib) (4.25.0)\nRequirement already satisfied: pyparsing>=2.2.1 in /databricks/python3/lib/python3.10/site-packages (from matplotlib) (3.0.9)\nRequirement already satisfied: python-dateutil>=2.7 in /databricks/python3/lib/python3.10/site-packages (from matplotlib) (2.8.2)\nRequirement already satisfied: cycler>=0.10 in /databricks/python3/lib/python3.10/site-packages (from matplotlib) (0.11.0)\nRequirement already satisfied: kiwisolver>=1.0.1 in /databricks/python3/lib/python3.10/site-packages (from matplotlib) (1.4.2)\nRequirement already satisfied: numpy>=1.17 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from matplotlib) (1.24.4)\nRequirement already satisfied: scikit-image in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from easyocr) (0.23.2)\nRequirement already satisfied: ninja in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from easyocr) (1.11.1.1)\nRequirement already satisfied: Shapely in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from easyocr) (2.0.4)\nRequirement already satisfied: python-bidi in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from easyocr) (0.4.2)\nRequirement already satisfied: scipy in /databricks/python3/lib/python3.10/site-packages (from easyocr) (1.9.1)\nRequirement already satisfied: pyclipper in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from easyocr) (1.3.0.post5)\nRequirement already satisfied: threadpoolctl>=2.0.0 in /databricks/python3/lib/python3.10/site-packages (from scikit-learn) (2.2.0)\nRequirement already satisfied: joblib>=1.0.0 in /databricks/python3/lib/python3.10/site-packages (from scikit-learn) (1.2.0)\nRequirement already satisfied: pytz>=2020.1 in /databricks/python3/lib/python3.10/site-packages (from pandas) (2022.1)\nRequirement already satisfied: jinja2>=2.10.3 in /databricks/python3/lib/python3.10/site-packages (from distributed) (2.11.3)\nRequirement already satisfied: sortedcontainers>=2.0.5 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from distributed) (2.4.0)\nRequirement already satisfied: zict>=3.0.0 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from distributed) (3.0.0)\nRequirement already satisfied: tornado>=6.0.4 in /databricks/python3/lib/python3.10/site-packages (from distributed) (6.1)\nRequirement already satisfied: psutil>=5.7.2 in /databricks/python3/lib/python3.10/site-packages (from distributed) (5.9.0)\nRequirement already satisfied: urllib3>=1.24.3 in /databricks/python3/lib/python3.10/site-packages (from distributed) (1.26.11)\nRequirement already satisfied: tblib>=1.6.0 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from distributed) (3.0.0)\nRequirement already satisfied: locket>=1.0.0 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from distributed) (1.0.0)\nRequirement already satisfied: msgpack>=1.0.0 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from distributed) (1.0.8)\nRequirement already satisfied: nvidia-cuda-runtime-cu12==12.1.105 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from torch) (12.1.105)\nRequirement already satisfied: nvidia-nccl-cu12==2.20.5 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from torch) (2.20.5)\nRequirement already satisfied: nvidia-cuda-cupti-cu12==12.1.105 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from torch) (12.1.105)\nRequirement already satisfied: nvidia-cublas-cu12==12.1.3.1 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from torch) (12.1.3.1)\nRequirement already satisfied: nvidia-cufft-cu12==11.0.2.54 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from torch) (11.0.2.54)\nRequirement already satisfied: nvidia-cusparse-cu12==12.1.0.106 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from torch) (12.1.0.106)\nRequirement already satisfied: sympy in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from torch) (1.12.1)\nRequirement already satisfied: nvidia-curand-cu12==10.3.2.106 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from torch) (10.3.2.106)\nRequirement already satisfied: triton==2.3.1 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from torch) (2.3.1)\nRequirement already satisfied: filelock in /usr/local/lib/python3.10/dist-packages (from torch) (3.12.2)\nRequirement already satisfied: nvidia-cuda-nvrtc-cu12==12.1.105 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from torch) (12.1.105)\nRequirement already satisfied: nvidia-cusolver-cu12==11.4.5.107 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from torch) (11.4.5.107)\nRequirement already satisfied: nvidia-cudnn-cu12==8.9.2.26 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from torch) (8.9.2.26)\nRequirement already satisfied: networkx in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from torch) (3.3)\nRequirement already satisfied: typing-extensions>=4.8.0 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from torch) (4.12.2)\nRequirement already satisfied: nvidia-nvtx-cu12==12.1.105 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from torch) (12.1.105)\nRequirement already satisfied: nvidia-nvjitlink-cu12 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from nvidia-cusolver-cu12==11.4.5.107->torch) (12.5.40)\nRequirement already satisfied: py4j==0.10.9.7 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from pyspark) (0.10.9.7)\nRequirement already satisfied: s3transfer<0.7.0,>=0.6.0 in /databricks/python3/lib/python3.10/site-packages (from boto3) (0.6.0)\nRequirement already satisfied: jmespath<2.0.0,>=0.7.1 in /databricks/python3/lib/python3.10/site-packages (from boto3) (0.10.0)\nRequirement already satisfied: botocore<1.28.0,>=1.27.28 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from boto3) (1.27.59)\nRequirement already satisfied: aiohttp!=4.0.0a0,!=4.0.0a1 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from s3fs) (3.9.5)\nRequirement already satisfied: aiobotocore~=2.4.2 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from s3fs) (2.4.2)\nRequirement already satisfied: aioitertools>=0.5.1 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from aiobotocore~=2.4.2->s3fs) (0.11.0)\nRequirement already satisfied: wrapt>=1.10.10 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from aiobotocore~=2.4.2->s3fs) (1.16.0)\nRequirement already satisfied: async-timeout<5.0,>=4.0 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs) (4.0.3)\nRequirement already satisfied: multidict<7.0,>=4.5 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs) (6.0.5)\nRequirement already satisfied: yarl<2.0,>=1.0 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs) (1.9.4)\nRequirement already satisfied: aiosignal>=1.1.2 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs) (1.3.1)\nRequirement already satisfied: attrs>=17.3.0 in /databricks/python3/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs) (21.4.0)\nRequirement already satisfied: frozenlist>=1.1.1 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs) (1.4.1)\nRequirement already satisfied: zipp>=0.5 in /usr/lib/python3/dist-packages (from importlib-metadata>=4.13.0->dask) (1.0.0)\nRequirement already satisfied: MarkupSafe>=0.23 in /databricks/python3/lib/python3.10/site-packages (from jinja2>=2.10.3->distributed) (2.0.1)\nRequirement already satisfied: six>=1.5 in /usr/lib/python3/dist-packages (from python-dateutil>=2.7->matplotlib) (1.16.0)\nRequirement already satisfied: cffi>=1.1.0 in /databricks/python3/lib/python3.10/site-packages (from cairocffi->cairosvg) (1.15.1)\nRequirement already satisfied: webencodings in /databricks/python3/lib/python3.10/site-packages (from cssselect2->cairosvg) (0.5.1)\nRequirement already satisfied: tifffile>=2022.8.12 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from scikit-image->easyocr) (2024.5.22)\nRequirement already satisfied: imageio>=2.33 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from scikit-image->easyocr) (2.34.1)\nRequirement already satisfied: lazy-loader>=0.4 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from scikit-image->easyocr) (0.4)\nRequirement already satisfied: mpmath<1.4.0,>=1.1.0 in /local_disk0/.ephemeral_nfs/envs/pythonEnv-b4d7ed6d-430a-4d97-ab0f-bb59bd09e258/lib/python3.10/site-packages (from sympy->torch) (1.3.0)\nRequirement already satisfied: pycparser in /databricks/python3/lib/python3.10/site-packages (from cffi>=1.1.0->cairocffi->cairosvg) (2.21)\nRequirement already satisfied: idna>=2.0 in /databricks/python3/lib/python3.10/site-packages (from yarl<2.0,>=1.0->aiohttp!=4.0.0a0,!=4.0.0a1->s3fs) (3.3)\n\u001B[43mNote: you may need to restart the kernel using dbutils.library.restartPython() to use updated packages.\u001B[0m\n"
     ]
    }
   ],
   "source": [
    "%pip install tqdm dask cairosvg pillow matplotlib easyocr scikit-learn opencv-python-headless pandas distributed torch torchvision torchaudio pyspark setuptools boto3 s3fs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "metadata": {
    "application/vnd.databricks.v1+cell": {
     "cellMetadata": {
      "byteLimit": 2048000,
      "rowLimit": 10000
     },
     "inputWidgets": {},
     "nuid": "f4b450ef-6c50-45f6-b4c1-cb537ec024ab",
     "showTitle": false,
     "title": ""
    }
   },
   "outputs": [
    {
     "output_type": "display_data",
     "data": {
      "text/plain": [
       "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m\n",
       "\u001B[0;31mExecutionError\u001B[0m                            Traceback (most recent call last)\n",
       "File \u001B[0;32m<command-4411400619119877>, line 2\u001B[0m\n",
       "\u001B[1;32m      1\u001B[0m \u001B[38;5;66;03m# Mount the S3 bucket\u001B[39;00m\n",
       "\u001B[0;32m----> 2\u001B[0m \u001B[43mdbutils\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mfs\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mmount\u001B[49m\u001B[43m(\u001B[49m\n",
       "\u001B[1;32m      3\u001B[0m \u001B[43m  \u001B[49m\u001B[43msource\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43m \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43ms3a://scisketch-dataset\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\n",
       "\u001B[1;32m      4\u001B[0m \u001B[43m  \u001B[49m\u001B[43mmount_point\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43m \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43m/mnt/scisketch\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\n",
       "\u001B[1;32m      5\u001B[0m \u001B[43m  \u001B[49m\u001B[43mextra_configs\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43m \u001B[49m\u001B[43m{\u001B[49m\n",
       "\u001B[1;32m      6\u001B[0m \u001B[43m      \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mfs.s3a.access.key\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m:\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mAKIA6GBMHS3ZLHCJN4FH\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\n",
       "\u001B[1;32m      7\u001B[0m \u001B[43m      \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mfs.s3a.secret.key\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m:\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mQVenBpsLWRAIhSPdj1TIwdC9XmE1CCWTFc0WAPzb\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\n",
       "\u001B[1;32m      8\u001B[0m \u001B[43m  \u001B[49m\u001B[43m}\u001B[49m\n",
       "\u001B[1;32m      9\u001B[0m \u001B[43m)\u001B[49m\n",
       "\n",
       "File \u001B[0;32m/databricks/python_shell/dbruntime/dbutils.py:362\u001B[0m, in \u001B[0;36mDBUtils.FSHandler.prettify_exception_message.<locals>.f_with_exception_handling\u001B[0;34m(*args, **kwargs)\u001B[0m\n",
       "\u001B[1;32m    360\u001B[0m exc\u001B[38;5;241m.\u001B[39m__context__ \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m\n",
       "\u001B[1;32m    361\u001B[0m exc\u001B[38;5;241m.\u001B[39m__cause__ \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m\n",
       "\u001B[0;32m--> 362\u001B[0m \u001B[38;5;28;01mraise\u001B[39;00m exc\n",
       "\n",
       "\u001B[0;31mExecutionError\u001B[0m: An error occurred while calling o397.mount.\n",
       ": java.rmi.RemoteException: java.lang.IllegalArgumentException: requirement failed: Directory already mounted: /mnt/scisketch; nested exception is: \n",
       "\tjava.lang.IllegalArgumentException: requirement failed: Directory already mounted: /mnt/scisketch\n",
       "\tat com.databricks.backend.daemon.data.client.DbfsClient.send0(DbfsClient.scala:135)\n",
       "\tat com.databricks.backend.daemon.data.client.DbfsClient.sendIdempotent(DbfsClient.scala:69)\n",
       "\tat com.databricks.backend.daemon.dbutils.DBUtilsCore.createOrUpdateMount(DBUtilsCore.scala:1053)\n",
       "\tat com.databricks.backend.daemon.dbutils.DBUtilsCore.$anonfun$mount$1(DBUtilsCore.scala:1079)\n",
       "\tat com.databricks.logging.UsageLogging.$anonfun$recordOperation$1(UsageLogging.scala:571)\n",
       "\tat com.databricks.logging.UsageLogging.executeThunkAndCaptureResultTags$1(UsageLogging.scala:667)\n",
       "\tat com.databricks.logging.UsageLogging.$anonfun$recordOperationWithResultTags$4(UsageLogging.scala:685)\n",
       "\tat com.databricks.logging.UsageLogging.$anonfun$withAttributionContext$1(UsageLogging.scala:426)\n",
       "\tat scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)\n",
       "\tat com.databricks.logging.AttributionContext$.withValue(AttributionContext.scala:196)\n",
       "\tat com.databricks.logging.UsageLogging.withAttributionContext(UsageLogging.scala:424)\n",
       "\tat com.databricks.logging.UsageLogging.withAttributionContext$(UsageLogging.scala:418)\n",
       "\tat com.databricks.backend.daemon.dbutils.FSUtils.withAttributionContext(DBUtilsCore.scala:71)\n",
       "\tat com.databricks.logging.UsageLogging.withAttributionTags(UsageLogging.scala:470)\n",
       "\tat com.databricks.logging.UsageLogging.withAttributionTags$(UsageLogging.scala:455)\n",
       "\tat com.databricks.backend.daemon.dbutils.FSUtils.withAttributionTags(DBUtilsCore.scala:71)\n",
       "\tat com.databricks.logging.UsageLogging.recordOperationWithResultTags(UsageLogging.scala:662)\n",
       "\tat com.databricks.logging.UsageLogging.recordOperationWithResultTags$(UsageLogging.scala:580)\n",
       "\tat com.databricks.backend.daemon.dbutils.FSUtils.recordOperationWithResultTags(DBUtilsCore.scala:71)\n",
       "\tat com.databricks.logging.UsageLogging.recordOperation(UsageLogging.scala:571)\n",
       "\tat com.databricks.logging.UsageLogging.recordOperation$(UsageLogging.scala:540)\n",
       "\tat com.databricks.backend.daemon.dbutils.FSUtils.recordOperation(DBUtilsCore.scala:71)\n",
       "\tat com.databricks.backend.daemon.dbutils.FSUtils.recordDbutilsFsOp(DBUtilsCore.scala:135)\n",
       "\tat com.databricks.backend.daemon.dbutils.DBUtilsCore.mount(DBUtilsCore.scala:1073)\n",
       "\tat sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n",
       "\tat sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n",
       "\tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n",
       "\tat java.lang.reflect.Method.invoke(Method.java:498)\n",
       "\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)\n",
       "\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:397)\n",
       "\tat py4j.Gateway.invoke(Gateway.java:306)\n",
       "\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)\n",
       "\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n",
       "\tat py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:199)\n",
       "\tat py4j.ClientServerConnection.run(ClientServerConnection.java:119)\n",
       "\tat java.lang.Thread.run(Thread.java:750)\n",
       "Caused by: java.lang.IllegalArgumentException: requirement failed: Directory already mounted: /mnt/scisketch\n",
       "\tat scala.Predef$.require(Predef.scala:281)\n",
       "\tat com.databricks.backend.daemon.data.server.DefaultMetadataManager.$anonfun$insertMount$1(MetadataManager.scala:705)\n",
       "\tat com.databricks.backend.daemon.data.server.DefaultMetadataManager.$anonfun$modifyAndVerify$2(MetadataManager.scala:1087)\n",
       "\tat scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)\n",
       "\tat com.databricks.backend.daemon.data.server.DefaultMetadataManager.withRetries(MetadataManager.scala:860)\n",
       "\tat com.databricks.backend.daemon.data.server.DefaultMetadataManager.modifyAndVerify(MetadataManager.scala:1076)\n",
       "\tat com.databricks.backend.daemon.data.server.DefaultMetadataManager.insertMount(MetadataManager.scala:713)\n",
       "\tat com.databricks.backend.daemon.data.server.handler.MountHandler.receive(MountHandler.scala:128)\n",
       "\tat com.databricks.backend.daemon.data.server.session.SessionContext.$anonfun$queryHandlers$1(SessionContext.scala:54)\n",
       "\tat com.databricks.backend.daemon.data.server.session.SessionContext.$anonfun$queryHandlers$1$adapted(SessionContext.scala:53)\n",
       "\tat scala.collection.immutable.List.foreach(List.scala:431)\n",
       "\tat com.databricks.backend.daemon.data.server.session.SessionContext.queryHandlers(SessionContext.scala:53)\n",
       "\tat com.databricks.backend.daemon.data.server.DbfsServerBackend$$anonfun$receive$4.$anonfun$applyOrElse$10(DbfsServerBackend.scala:472)\n",
       "\tat com.databricks.logging.AttributionContextTracing.$anonfun$withAttributionContext$1(AttributionContextTracing.scala:48)\n",
       "\tat com.databricks.logging.AttributionContext$.$anonfun$withValue$1(AttributionContext.scala:253)\n",
       "\tat scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)\n",
       "\tat com.databricks.logging.AttributionContext$.withValue(AttributionContext.scala:249)\n",
       "\tat com.databricks.logging.AttributionContextTracing.withAttributionContext(AttributionContextTracing.scala:46)\n",
       "\tat com.databricks.logging.AttributionContextTracing.withAttributionContext$(AttributionContextTracing.scala:43)\n",
       "\tat com.databricks.rpc.ServerBackend.withAttributionContext(ServerBackend.scala:22)\n",
       "\tat com.databricks.logging.AttributionContextTracing.withAttributionTags(AttributionContextTracing.scala:95)\n",
       "\tat com.databricks.logging.AttributionContextTracing.withAttributionTags$(AttributionContextTracing.scala:76)\n",
       "\tat com.databricks.rpc.ServerBackend.withAttributionTags(ServerBackend.scala:22)\n",
       "\tat com.databricks.backend.daemon.data.server.DbfsServerBackend$$anonfun$receive$4.applyOrElse(DbfsServerBackend.scala:472)\n",
       "\tat com.databricks.backend.daemon.data.server.DbfsServerBackend$$anonfun$receive$4.applyOrElse(DbfsServerBackend.scala:365)\n",
       "\tat com.databricks.rpc.ServerBackend.$anonfun$internalReceive0$2(ServerBackend.scala:174)\n",
       "\tat com.databricks.rpc.ServerBackend$$anonfun$commonReceive$1.applyOrElse(ServerBackend.scala:200)\n",
       "\tat com.databricks.rpc.ServerBackend$$anonfun$commonReceive$1.applyOrElse(ServerBackend.scala:200)\n",
       "\tat com.databricks.rpc.ServerBackend.internalReceive0(ServerBackend.scala:171)\n",
       "\tat com.databricks.rpc.ServerBackend.$anonfun$internalReceive$1(ServerBackend.scala:147)\n",
       "\tat com.databricks.logging.UsageLogging.$anonfun$recordOperation$1(UsageLogging.scala:532)\n",
       "\tat com.databricks.logging.UsageLogging.executeThunkAndCaptureResultTags$1(UsageLogging.scala:636)\n",
       "\tat com.databricks.logging.UsageLogging.$anonfun$recordOperationWithResultTags$4(UsageLogging.scala:654)\n",
       "\tat com.databricks.logging.AttributionContextTracing.$anonfun$withAttributionContext$1(AttributionContextTracing.scala:48)\n",
       "\tat com.databricks.logging.AttributionContext$.$anonfun$withValue$1(AttributionContext.scala:253)\n",
       "\tat scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)\n",
       "\tat com.databricks.logging.AttributionContext$.withValue(AttributionContext.scala:249)\n",
       "\tat com.databricks.logging.AttributionContextTracing.withAttributionContext(AttributionContextTracing.scala:46)\n",
       "\tat com.databricks.logging.AttributionContextTracing.withAttributionContext$(AttributionContextTracing.scala:43)\n",
       "\tat com.databricks.rpc.ServerBackend.withAttributionContext(ServerBackend.scala:22)\n",
       "\tat com.databricks.logging.AttributionContextTracing.withAttributionTags(AttributionContextTracing.scala:95)\n",
       "\tat com.databricks.logging.AttributionContextTracing.withAttributionTags$(AttributionContextTracing.scala:76)\n",
       "\tat com.databricks.rpc.ServerBackend.withAttributionTags(ServerBackend.scala:22)\n",
       "\tat com.databricks.logging.UsageLogging.recordOperationWithResultTags(UsageLogging.scala:631)\n",
       "\tat com.databricks.logging.UsageLogging.recordOperationWithResultTags$(UsageLogging.scala:541)\n",
       "\tat com.databricks.rpc.ServerBackend.recordOperationWithResultTags(ServerBackend.scala:22)\n",
       "\tat com.databricks.logging.UsageLogging.recordOperation(UsageLogging.scala:533)\n",
       "\tat com.databricks.logging.UsageLogging.recordOperation$(UsageLogging.scala:501)\n",
       "\tat com.databricks.rpc.ServerBackend.recordOperation(ServerBackend.scala:22)\n",
       "\tat com.databricks.rpc.ServerBackend.internalReceive(ServerBackend.scala:146)\n",
       "\tat com.databricks.rpc.JettyServer$RequestManager.handleRPC(JettyServer.scala:1021)\n",
       "\tat com.databricks.rpc.JettyServer$RequestManager.handleRequestAndRespond(JettyServer.scala:942)\n",
       "\tat com.databricks.rpc.JettyServer$RequestManager.$anonfun$handleHttp$6(JettyServer.scala:546)\n",
       "\tat com.databricks.rpc.JettyServer$RequestManager.$anonfun$handleHttp$6$adapted(JettyServer.scala:515)\n",
       "\tat com.databricks.logging.activity.ActivityContextFactory$.$anonfun$withActivityInternal$6(ActivityContextFactory.scala:549)\n",
       "\tat com.databricks.logging.AttributionContextTracing.$anonfun$withAttributionContext$1(AttributionContextTracing.scala:48)\n",
       "\tat com.databricks.logging.AttributionContext$.$anonfun$withValue$1(AttributionContext.scala:253)\n",
       "\tat scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)\n",
       "\tat com.databricks.logging.AttributionContext$.withValue(AttributionContext.scala:249)\n",
       "\tat com.databricks.logging.AttributionContextTracing.withAttributionContext(AttributionContextTracing.scala:46)\n",
       "\tat com.databricks.logging.AttributionContextTracing.withAttributionContext$(AttributionContextTracing.scala:43)\n",
       "\tat com.databricks.logging.activity.ActivityContextFactory$.withAttributionContext(ActivityContextFactory.scala:57)\n",
       "\tat com.databricks.logging.activity.ActivityContextFactory$.$anonfun$withActivityInternal$3(ActivityContextFactory.scala:549)\n",
       "\tat com.databricks.context.integrity.IntegrityCheckContext$ThreadLocalStorage$.withValue(IntegrityCheckContext.scala:73)\n",
       "\tat com.databricks.logging.activity.ActivityContextFactory$.withActivityInternal(ActivityContextFactory.scala:527)\n",
       "\tat com.databricks.logging.activity.ActivityContextFactory$.withServiceRequestActivity(ActivityContextFactory.scala:178)\n",
       "\tat com.databricks.rpc.JettyServer$RequestManager.handleHttp(JettyServer.scala:515)\n",
       "\tat com.databricks.rpc.JettyServer$RequestManager.doPost(JettyServer.scala:405)\n",
       "\tat javax.servlet.http.HttpServlet.service(HttpServlet.java:665)\n",
       "\tat com.databricks.rpc.HttpServletWithPatch.service(HttpServletWithPatch.scala:33)\n",
       "\tat javax.servlet.http.HttpServlet.service(HttpServlet.java:750)\n",
       "\tat org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)\n",
       "\tat org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:554)\n",
       "\tat org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:190)\n",
       "\tat org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505)\n",
       "\tat org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)\n",
       "\tat org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)\n",
       "\tat org.eclipse.jetty.server.Server.handle(Server.java:516)\n",
       "\tat org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)\n",
       "\tat org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)\n",
       "\tat org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)\n",
       "\tat org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)\n",
       "\tat org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)\n",
       "\tat org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)\n",
       "\tat org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)\n",
       "\tat org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)\n",
       "\tat org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)\n",
       "\tat org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)\n",
       "\tat org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)\n",
       "\tat org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409)\n",
       "\tat com.databricks.rpc.InstrumentedQueuedThreadPool$$anon$1.$anonfun$run$2(InstrumentedQueuedThreadPool.scala:107)\n",
       "\tat scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)\n",
       "\tat com.databricks.logging.AttributionContextTracing.$anonfun$withAttributionContext$1(AttributionContextTracing.scala:48)\n",
       "\tat com.databricks.logging.AttributionContext$.$anonfun$withValue$1(AttributionContext.scala:253)\n",
       "\tat scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)\n",
       "\tat com.databricks.logging.AttributionContext$.withValue(AttributionContext.scala:249)\n",
       "\tat com.databricks.logging.AttributionContextTracing.withAttributionContext(AttributionContextTracing.scala:46)\n",
       "\tat com.databricks.logging.AttributionContextTracing.withAttributionContext$(AttributionContextTracing.scala:43)\n",
       "\tat com.databricks.rpc.InstrumentedQueuedThreadPool.withAttributionContext(InstrumentedQueuedThreadPool.scala:46)\n",
       "\tat com.databricks.rpc.InstrumentedQueuedThreadPool$$anon$1.$anonfun$run$1(InstrumentedQueuedThreadPool.scala:107)\n",
       "\tat scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)\n",
       "\tat com.databricks.instrumentation.QueuedThreadPoolInstrumenter.trackActiveThreads(QueuedThreadPoolInstrumenter.scala:110)\n",
       "\tat com.databricks.instrumentation.QueuedThreadPoolInstrumenter.trackActiveThreads$(QueuedThreadPoolInstrumenter.scala:107)\n",
       "\tat com.databricks.rpc.InstrumentedQueuedThreadPool.trackActiveThreads(InstrumentedQueuedThreadPool.scala:46)\n",
       "\tat com.databricks.rpc.InstrumentedQueuedThreadPool$$anon$1.run(InstrumentedQueuedThreadPool.scala:89)\n",
       "\tat org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)\n",
       "\tat org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)\n",
       "\tat java.lang.Thread.run(Thread.java:840)\n"
      ]
     },
     "metadata": {
      "application/vnd.databricks.v1+output": {
       "arguments": {},
       "data": "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m\n\u001B[0;31mExecutionError\u001B[0m                            Traceback (most recent call last)\nFile \u001B[0;32m<command-4411400619119877>, line 2\u001B[0m\n\u001B[1;32m      1\u001B[0m \u001B[38;5;66;03m# Mount the S3 bucket\u001B[39;00m\n\u001B[0;32m----> 2\u001B[0m \u001B[43mdbutils\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mfs\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mmount\u001B[49m\u001B[43m(\u001B[49m\n\u001B[1;32m      3\u001B[0m \u001B[43m  \u001B[49m\u001B[43msource\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43m \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43ms3a://scisketch-dataset\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\n\u001B[1;32m      4\u001B[0m \u001B[43m  \u001B[49m\u001B[43mmount_point\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43m \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43m/mnt/scisketch\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\n\u001B[1;32m      5\u001B[0m \u001B[43m  \u001B[49m\u001B[43mextra_configs\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43m \u001B[49m\u001B[43m{\u001B[49m\n\u001B[1;32m      6\u001B[0m \u001B[43m      \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mfs.s3a.access.key\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m:\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mAKIA6GBMHS3ZLHCJN4FH\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\n\u001B[1;32m      7\u001B[0m \u001B[43m      \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mfs.s3a.secret.key\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m:\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mQVenBpsLWRAIhSPdj1TIwdC9XmE1CCWTFc0WAPzb\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\n\u001B[1;32m      8\u001B[0m \u001B[43m  \u001B[49m\u001B[43m}\u001B[49m\n\u001B[1;32m      9\u001B[0m \u001B[43m)\u001B[49m\n\nFile \u001B[0;32m/databricks/python_shell/dbruntime/dbutils.py:362\u001B[0m, in \u001B[0;36mDBUtils.FSHandler.prettify_exception_message.<locals>.f_with_exception_handling\u001B[0;34m(*args, **kwargs)\u001B[0m\n\u001B[1;32m    360\u001B[0m exc\u001B[38;5;241m.\u001B[39m__context__ \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m\n\u001B[1;32m    361\u001B[0m exc\u001B[38;5;241m.\u001B[39m__cause__ \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m\n\u001B[0;32m--> 362\u001B[0m \u001B[38;5;28;01mraise\u001B[39;00m exc\n\n\u001B[0;31mExecutionError\u001B[0m: An error occurred while calling o397.mount.\n: java.rmi.RemoteException: java.lang.IllegalArgumentException: requirement failed: Directory already mounted: /mnt/scisketch; nested exception is: \n\tjava.lang.IllegalArgumentException: requirement failed: Directory already mounted: /mnt/scisketch\n\tat com.databricks.backend.daemon.data.client.DbfsClient.send0(DbfsClient.scala:135)\n\tat com.databricks.backend.daemon.data.client.DbfsClient.sendIdempotent(DbfsClient.scala:69)\n\tat com.databricks.backend.daemon.dbutils.DBUtilsCore.createOrUpdateMount(DBUtilsCore.scala:1053)\n\tat com.databricks.backend.daemon.dbutils.DBUtilsCore.$anonfun$mount$1(DBUtilsCore.scala:1079)\n\tat com.databricks.logging.UsageLogging.$anonfun$recordOperation$1(UsageLogging.scala:571)\n\tat com.databricks.logging.UsageLogging.executeThunkAndCaptureResultTags$1(UsageLogging.scala:667)\n\tat com.databricks.logging.UsageLogging.$anonfun$recordOperationWithResultTags$4(UsageLogging.scala:685)\n\tat com.databricks.logging.UsageLogging.$anonfun$withAttributionContext$1(UsageLogging.scala:426)\n\tat scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)\n\tat com.databricks.logging.AttributionContext$.withValue(AttributionContext.scala:196)\n\tat com.databricks.logging.UsageLogging.withAttributionContext(UsageLogging.scala:424)\n\tat com.databricks.logging.UsageLogging.withAttributionContext$(UsageLogging.scala:418)\n\tat com.databricks.backend.daemon.dbutils.FSUtils.withAttributionContext(DBUtilsCore.scala:71)\n\tat com.databricks.logging.UsageLogging.withAttributionTags(UsageLogging.scala:470)\n\tat com.databricks.logging.UsageLogging.withAttributionTags$(UsageLogging.scala:455)\n\tat com.databricks.backend.daemon.dbutils.FSUtils.withAttributionTags(DBUtilsCore.scala:71)\n\tat com.databricks.logging.UsageLogging.recordOperationWithResultTags(UsageLogging.scala:662)\n\tat com.databricks.logging.UsageLogging.recordOperationWithResultTags$(UsageLogging.scala:580)\n\tat com.databricks.backend.daemon.dbutils.FSUtils.recordOperationWithResultTags(DBUtilsCore.scala:71)\n\tat com.databricks.logging.UsageLogging.recordOperation(UsageLogging.scala:571)\n\tat com.databricks.logging.UsageLogging.recordOperation$(UsageLogging.scala:540)\n\tat com.databricks.backend.daemon.dbutils.FSUtils.recordOperation(DBUtilsCore.scala:71)\n\tat com.databricks.backend.daemon.dbutils.FSUtils.recordDbutilsFsOp(DBUtilsCore.scala:135)\n\tat com.databricks.backend.daemon.dbutils.DBUtilsCore.mount(DBUtilsCore.scala:1073)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.lang.reflect.Method.invoke(Method.java:498)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:397)\n\tat py4j.Gateway.invoke(Gateway.java:306)\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n\tat py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:199)\n\tat py4j.ClientServerConnection.run(ClientServerConnection.java:119)\n\tat java.lang.Thread.run(Thread.java:750)\nCaused by: java.lang.IllegalArgumentException: requirement failed: Directory already mounted: /mnt/scisketch\n\tat scala.Predef$.require(Predef.scala:281)\n\tat com.databricks.backend.daemon.data.server.DefaultMetadataManager.$anonfun$insertMount$1(MetadataManager.scala:705)\n\tat com.databricks.backend.daemon.data.server.DefaultMetadataManager.$anonfun$modifyAndVerify$2(MetadataManager.scala:1087)\n\tat scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)\n\tat com.databricks.backend.daemon.data.server.DefaultMetadataManager.withRetries(MetadataManager.scala:860)\n\tat com.databricks.backend.daemon.data.server.DefaultMetadataManager.modifyAndVerify(MetadataManager.scala:1076)\n\tat com.databricks.backend.daemon.data.server.DefaultMetadataManager.insertMount(MetadataManager.scala:713)\n\tat com.databricks.backend.daemon.data.server.handler.MountHandler.receive(MountHandler.scala:128)\n\tat com.databricks.backend.daemon.data.server.session.SessionContext.$anonfun$queryHandlers$1(SessionContext.scala:54)\n\tat com.databricks.backend.daemon.data.server.session.SessionContext.$anonfun$queryHandlers$1$adapted(SessionContext.scala:53)\n\tat scala.collection.immutable.List.foreach(List.scala:431)\n\tat com.databricks.backend.daemon.data.server.session.SessionContext.queryHandlers(SessionContext.scala:53)\n\tat com.databricks.backend.daemon.data.server.DbfsServerBackend$$anonfun$receive$4.$anonfun$applyOrElse$10(DbfsServerBackend.scala:472)\n\tat com.databricks.logging.AttributionContextTracing.$anonfun$withAttributionContext$1(AttributionContextTracing.scala:48)\n\tat com.databricks.logging.AttributionContext$.$anonfun$withValue$1(AttributionContext.scala:253)\n\tat scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)\n\tat com.databricks.logging.AttributionContext$.withValue(AttributionContext.scala:249)\n\tat com.databricks.logging.AttributionContextTracing.withAttributionContext(AttributionContextTracing.scala:46)\n\tat com.databricks.logging.AttributionContextTracing.withAttributionContext$(AttributionContextTracing.scala:43)\n\tat com.databricks.rpc.ServerBackend.withAttributionContext(ServerBackend.scala:22)\n\tat com.databricks.logging.AttributionContextTracing.withAttributionTags(AttributionContextTracing.scala:95)\n\tat com.databricks.logging.AttributionContextTracing.withAttributionTags$(AttributionContextTracing.scala:76)\n\tat com.databricks.rpc.ServerBackend.withAttributionTags(ServerBackend.scala:22)\n\tat com.databricks.backend.daemon.data.server.DbfsServerBackend$$anonfun$receive$4.applyOrElse(DbfsServerBackend.scala:472)\n\tat com.databricks.backend.daemon.data.server.DbfsServerBackend$$anonfun$receive$4.applyOrElse(DbfsServerBackend.scala:365)\n\tat com.databricks.rpc.ServerBackend.$anonfun$internalReceive0$2(ServerBackend.scala:174)\n\tat com.databricks.rpc.ServerBackend$$anonfun$commonReceive$1.applyOrElse(ServerBackend.scala:200)\n\tat com.databricks.rpc.ServerBackend$$anonfun$commonReceive$1.applyOrElse(ServerBackend.scala:200)\n\tat com.databricks.rpc.ServerBackend.internalReceive0(ServerBackend.scala:171)\n\tat com.databricks.rpc.ServerBackend.$anonfun$internalReceive$1(ServerBackend.scala:147)\n\tat com.databricks.logging.UsageLogging.$anonfun$recordOperation$1(UsageLogging.scala:532)\n\tat com.databricks.logging.UsageLogging.executeThunkAndCaptureResultTags$1(UsageLogging.scala:636)\n\tat com.databricks.logging.UsageLogging.$anonfun$recordOperationWithResultTags$4(UsageLogging.scala:654)\n\tat com.databricks.logging.AttributionContextTracing.$anonfun$withAttributionContext$1(AttributionContextTracing.scala:48)\n\tat com.databricks.logging.AttributionContext$.$anonfun$withValue$1(AttributionContext.scala:253)\n\tat scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)\n\tat com.databricks.logging.AttributionContext$.withValue(AttributionContext.scala:249)\n\tat com.databricks.logging.AttributionContextTracing.withAttributionContext(AttributionContextTracing.scala:46)\n\tat com.databricks.logging.AttributionContextTracing.withAttributionContext$(AttributionContextTracing.scala:43)\n\tat com.databricks.rpc.ServerBackend.withAttributionContext(ServerBackend.scala:22)\n\tat com.databricks.logging.AttributionContextTracing.withAttributionTags(AttributionContextTracing.scala:95)\n\tat com.databricks.logging.AttributionContextTracing.withAttributionTags$(AttributionContextTracing.scala:76)\n\tat com.databricks.rpc.ServerBackend.withAttributionTags(ServerBackend.scala:22)\n\tat com.databricks.logging.UsageLogging.recordOperationWithResultTags(UsageLogging.scala:631)\n\tat com.databricks.logging.UsageLogging.recordOperationWithResultTags$(UsageLogging.scala:541)\n\tat com.databricks.rpc.ServerBackend.recordOperationWithResultTags(ServerBackend.scala:22)\n\tat com.databricks.logging.UsageLogging.recordOperation(UsageLogging.scala:533)\n\tat com.databricks.logging.UsageLogging.recordOperation$(UsageLogging.scala:501)\n\tat com.databricks.rpc.ServerBackend.recordOperation(ServerBackend.scala:22)\n\tat com.databricks.rpc.ServerBackend.internalReceive(ServerBackend.scala:146)\n\tat com.databricks.rpc.JettyServer$RequestManager.handleRPC(JettyServer.scala:1021)\n\tat com.databricks.rpc.JettyServer$RequestManager.handleRequestAndRespond(JettyServer.scala:942)\n\tat com.databricks.rpc.JettyServer$RequestManager.$anonfun$handleHttp$6(JettyServer.scala:546)\n\tat com.databricks.rpc.JettyServer$RequestManager.$anonfun$handleHttp$6$adapted(JettyServer.scala:515)\n\tat com.databricks.logging.activity.ActivityContextFactory$.$anonfun$withActivityInternal$6(ActivityContextFactory.scala:549)\n\tat com.databricks.logging.AttributionContextTracing.$anonfun$withAttributionContext$1(AttributionContextTracing.scala:48)\n\tat com.databricks.logging.AttributionContext$.$anonfun$withValue$1(AttributionContext.scala:253)\n\tat scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)\n\tat com.databricks.logging.AttributionContext$.withValue(AttributionContext.scala:249)\n\tat com.databricks.logging.AttributionContextTracing.withAttributionContext(AttributionContextTracing.scala:46)\n\tat com.databricks.logging.AttributionContextTracing.withAttributionContext$(AttributionContextTracing.scala:43)\n\tat com.databricks.logging.activity.ActivityContextFactory$.withAttributionContext(ActivityContextFactory.scala:57)\n\tat com.databricks.logging.activity.ActivityContextFactory$.$anonfun$withActivityInternal$3(ActivityContextFactory.scala:549)\n\tat com.databricks.context.integrity.IntegrityCheckContext$ThreadLocalStorage$.withValue(IntegrityCheckContext.scala:73)\n\tat com.databricks.logging.activity.ActivityContextFactory$.withActivityInternal(ActivityContextFactory.scala:527)\n\tat com.databricks.logging.activity.ActivityContextFactory$.withServiceRequestActivity(ActivityContextFactory.scala:178)\n\tat com.databricks.rpc.JettyServer$RequestManager.handleHttp(JettyServer.scala:515)\n\tat com.databricks.rpc.JettyServer$RequestManager.doPost(JettyServer.scala:405)\n\tat javax.servlet.http.HttpServlet.service(HttpServlet.java:665)\n\tat com.databricks.rpc.HttpServletWithPatch.service(HttpServletWithPatch.scala:33)\n\tat javax.servlet.http.HttpServlet.service(HttpServlet.java:750)\n\tat org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)\n\tat org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:554)\n\tat org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:190)\n\tat org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505)\n\tat org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)\n\tat org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)\n\tat org.eclipse.jetty.server.Server.handle(Server.java:516)\n\tat org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)\n\tat org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)\n\tat org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)\n\tat org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)\n\tat org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)\n\tat org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)\n\tat org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)\n\tat org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)\n\tat org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)\n\tat org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)\n\tat org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)\n\tat org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409)\n\tat com.databricks.rpc.InstrumentedQueuedThreadPool$$anon$1.$anonfun$run$2(InstrumentedQueuedThreadPool.scala:107)\n\tat scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)\n\tat com.databricks.logging.AttributionContextTracing.$anonfun$withAttributionContext$1(AttributionContextTracing.scala:48)\n\tat com.databricks.logging.AttributionContext$.$anonfun$withValue$1(AttributionContext.scala:253)\n\tat scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)\n\tat com.databricks.logging.AttributionContext$.withValue(AttributionContext.scala:249)\n\tat com.databricks.logging.AttributionContextTracing.withAttributionContext(AttributionContextTracing.scala:46)\n\tat com.databricks.logging.AttributionContextTracing.withAttributionContext$(AttributionContextTracing.scala:43)\n\tat com.databricks.rpc.InstrumentedQueuedThreadPool.withAttributionContext(InstrumentedQueuedThreadPool.scala:46)\n\tat com.databricks.rpc.InstrumentedQueuedThreadPool$$anon$1.$anonfun$run$1(InstrumentedQueuedThreadPool.scala:107)\n\tat scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)\n\tat com.databricks.instrumentation.QueuedThreadPoolInstrumenter.trackActiveThreads(QueuedThreadPoolInstrumenter.scala:110)\n\tat com.databricks.instrumentation.QueuedThreadPoolInstrumenter.trackActiveThreads$(QueuedThreadPoolInstrumenter.scala:107)\n\tat com.databricks.rpc.InstrumentedQueuedThreadPool.trackActiveThreads(InstrumentedQueuedThreadPool.scala:46)\n\tat com.databricks.rpc.InstrumentedQueuedThreadPool$$anon$1.run(InstrumentedQueuedThreadPool.scala:89)\n\tat org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)\n\tat org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)\n\tat java.lang.Thread.run(Thread.java:840)\n",
       "errorSummary": "java.rmi.RemoteException: java.lang.IllegalArgumentException: requirement failed: Directory already mounted: /mnt/scisketch; nested exception is: ",
       "errorTraceType": "ansi",
       "metadata": {},
       "type": "ipynbError"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Mount the S3 bucket\n",
    "dbutils.fs.mount(\n",
    "  source = \"s3a://scisketch-dataset\",\n",
    "  mount_point = \"/mnt/scisketch\",\n",
    "  extra_configs = {\n",
    "      \"fs.s3a.access.key\": \"AKIA6GBMHS3ZLHCJN4FH\",\n",
    "      \"fs.s3a.secret.key\": \"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\"\n",
    "  }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "metadata": {
    "application/vnd.databricks.v1+cell": {
     "cellMetadata": {
      "byteLimit": 2048000,
      "rowLimit": 10000
     },
     "inputWidgets": {},
     "nuid": "e16c8c63-ebf7-4ea9-aff2-f5944c943e89",
     "showTitle": false,
     "title": ""
    }
   },
   "outputs": [],
   "source": [
    "import boto3\n",
    "import easyocr\n",
    "import cv2\n",
    "import numpy as np\n",
    "from sklearn.cluster import DBSCAN\n",
    "import torch\n",
    "import io\n",
    "from PIL import Image\n",
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv('s3a://scisketch-dataset/Entries_cleaned.csv')\n",
    "\n",
    "def process_image_path(image_path):\n",
    "    base_path = 's3a://scisketch-dataset/Graphical-Abstracts'\n",
    "    # Check if '/Images' is in the path\n",
    "    if '/Graphical Abstracts' in image_path:\n",
    "        # Split the path at '/Images' and take the part after it\n",
    "        processed_path = image_path.split('/Graphical Abstracts', 1)[-1]\n",
    "    else:\n",
    "        # If '/Images' is not in the path, use the original path\n",
    "        processed_path = image_path\n",
    "    # Concatenate the base path with the processed path\n",
    "    return base_path + processed_path\n",
    "\n",
    "df['Image'] = df['Image'].apply(process_image_path)\n",
    "\n",
    "# Drop the last row as it will have NaN in 'saved_image_path' after shifting\n",
    "df.dropna(subset=['Image'], inplace=True)\n",
    "\n",
    "# Reset the index after dropping rows\n",
    "df.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "metadata": {
    "application/vnd.databricks.v1+cell": {
     "cellMetadata": {
      "byteLimit": 2048000,
      "rowLimit": 10000
     },
     "inputWidgets": {},
     "nuid": "885b7c54-d5ed-42ba-9feb-d43c2091037e",
     "showTitle": false,
     "title": ""
    }
   },
   "outputs": [],
   "source": [
    "from pyspark.sql import SparkSession\n",
    "spark = SparkSession.builder.appName(\"ImageProcessingApp\").getOrCreate()\n",
    "spark_df = spark.createDataFrame(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "metadata": {
    "application/vnd.databricks.v1+cell": {
     "cellMetadata": {
      "byteLimit": 2048000,
      "rowLimit": 10000
     },
     "inputWidgets": {},
     "nuid": "3ea85967-1c80-4afb-b187-322e8ea42a2d",
     "showTitle": false,
     "title": ""
    }
   },
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File Name</th>\n",
       "      <th>DOI_urls</th>\n",
       "      <th>Abstract</th>\n",
       "      <th>DOI</th>\n",
       "      <th>Image</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ScienceDirect_citations_1715541681970.txt</td>\n",
       "      <td>https://doi.org/10.1016/j.molcel.2011.07.023.</td>\n",
       "      <td>In Escherichia coli, RNA degradation often begins with conversion of the 5′-terminal triphosphate to a monophosphate, creating a better substrate for internal cleavage by RNase E. Remarkably, no homolog of this key endonuclease is present in many bacterial species, such as Bacillus subtilis and various pathogens. Here, we report that the degradation of primary transcripts in B. subtilis can nevertheless be triggered by an analogous process to generate a short-lived, monophosphorylated intermediate. Like its E. coli counterpart, the B. subtilis RNA pyrophosphohydrolase that catalyzes this event is a Nudix protein that prefers unpaired 5′ ends. However, in B. subtilis, this modification exposes transcripts to rapid 5′ exonucleolytic degradation by RNase J, which is absent in E. coli but present in most bacteria lacking RNase E. This pathway, which closely resembles the mechanism by which deadenylated mRNA is degraded in eukaryotic cells, explains the stabilizing influence of 5′-terminal stem-loops in such bacteria.</td>\n",
       "      <td>10.1016/j.molcel.2011.07.023</td>\n",
       "      <td>s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.molcel.2011.07.023.jpg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ScienceDirect_citations_1715541681970.txt</td>\n",
       "      <td>https://doi.org/10.1016/j.molcel.2011.07.033.</td>\n",
       "      <td>The Drosophila ecdysone receptor (EcR/Usp) is thought to activate or repress gene transcription depending on the presence or absence, respectively, of the hormone ecdysone. Unexpectedly, we found an alternative mechanism at work in salivary glands during the ecdysone-dependent transition from larvae to pupae. In the absense of ecdysone, both ecdysone receptor subunits localize to the cytoplasm, and the heme-binding nuclear receptor E75A replaces EcR/Usp at common target sequences in several genes. During the larval-pupal transition, a switch from gene activation by EcR/Usp to gene repression by E75A is triggered by a decrease in ecdysone concentration and by direct repression of the EcR gene by E75A. Additional control is provided by developmentally timed modulation of E75A activity by NO, which inhibits recruitment of the corepressor SMRTER. These results suggest a mechanism for sequential modulation of gene expression during development by competing nuclear receptors and their effector molecules, ecdysone and NO.</td>\n",
       "      <td>10.1016/j.molcel.2011.07.033</td>\n",
       "      <td>s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.molcel.2011.07.033.jpg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ScienceDirect_citations_1715541681970.txt</td>\n",
       "      <td>https://doi.org/10.1016/j.ccr.2011.11.003.</td>\n",
       "      <td>Tumor immune surveillance and cancer immunotherapies are thought to depend on the intratumoral infiltration of activated CD8+ T cells. Intratumoral CD8+ T cells are rare and lack activity. IL-10 is thought to contribute to the underlying immune suppressive microenvironment. Defying those expectations we demonstrate that IL-10 induces several essential mechanisms for effective antitumor immune surveillance: infiltration and activation of intratumoral tumor-specific cytotoxic CD8+ T cells, expression of the Th1 cytokine interferon-γ (IFNγ) and granzymes in CD8+ T cells, and intratumoral antigen presentation molecules. Consequently, tumor immune surveillance is weakened in mice deficient for IL-10 whereas transgenic overexpression of IL-10 protects mice from carcinogenesis. Treatment with pegylated IL-10 restores tumor-specific intratumoral CD8+ T cell function and controls tumor growth.</td>\n",
       "      <td>10.1016/j.ccr.2011.11.003</td>\n",
       "      <td>s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.ccr.2011.11.003.jpg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ScienceDirect_citations_1715541681970.txt</td>\n",
       "      <td>https://doi.org/10.1016/j.molcel.2011.06.013.</td>\n",
       "      <td>MicroRNAs (miRNAs) regulate gene expression through translation repression and mRNA destabilization. However, the molecular mechanisms of miRNA silencing are still not well defined. Using a genetic screen in mouse embryonic stem (ES) cells, we identify mammalian hyperplastic discs protein EDD, a known E3 ubiquitin ligase, as a key component of the miRNA silencing pathway. ES cells deficient for EDD are defective in miRNA function and exhibit growth defects. We demonstrate that E3 ubiquitin ligase activity is dispensable for EDD function in miRNA silencing. Instead, EDD interacts with GW182 family proteins in the Argonaute-miRNA complexes. The PABC domain of EDD is essential for its silencing function. Through the PABC domain, EDD participates in miRNA silencing by recruiting downstream effectors. Among the PABC-interactors, DDX6 and Tob1/2 are both required and sufficient for silencing mRNA targets. Taken together, these data demonstrate a critical function for EDD in miRNA silencing.</td>\n",
       "      <td>10.1016/j.molcel.2011.06.013</td>\n",
       "      <td>s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.molcel.2011.06.013.jpg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ScienceDirect_citations_1715541681970.txt</td>\n",
       "      <td>https://doi.org/10.1016/j.molcel.2011.05.036.</td>\n",
       "      <td>The mechanisms by which ubiquitin ligases are regulated remain poorly understood. Here we describe a series of molecular events that coordinately regulate CHIP, a neuroprotective E3 implicated in protein quality control. Through their opposing activities, the initiator E2, Ube2w, and the specialized deubiquitinating enzyme (DUB), ataxin-3, participate in initiating, regulating, and terminating the CHIP ubiquitination cycle. Monoubiquitination of CHIP by Ube2w stabilizes the interaction between CHIP and ataxin-3, which through its DUB activity limits the length of chains attached to CHIP substrates. Upon completion of substrate ubiquitination, ataxin-3 deubiquitinates CHIP, effectively terminating the reaction. Our results suggest that functional pairing of E3s with ataxin-3 or similar DUBs represents an important point of regulation in ubiquitin-dependent protein quality control. In addition, the results shed light on disease pathogenesis in SCA3, a neurodegenerative disorder caused by polyglutamine expansion in ataxin-3.</td>\n",
       "      <td>10.1016/j.molcel.2011.05.036</td>\n",
       "      <td>s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.molcel.2011.05.036.jpg</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   File Name  ...                                                                         Image\n",
       "0  ScienceDirect_citations_1715541681970.txt  ...  s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.molcel.2011.07.023.jpg\n",
       "1  ScienceDirect_citations_1715541681970.txt  ...  s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.molcel.2011.07.033.jpg\n",
       "2  ScienceDirect_citations_1715541681970.txt  ...     s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.ccr.2011.11.003.jpg\n",
       "3  ScienceDirect_citations_1715541681970.txt  ...  s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.molcel.2011.06.013.jpg\n",
       "4  ScienceDirect_citations_1715541681970.txt  ...  s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.molcel.2011.05.036.jpg\n",
       "\n",
       "[5 rows x 5 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option('display.max_colwidth', None)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "metadata": {
    "application/vnd.databricks.v1+cell": {
     "cellMetadata": {
      "byteLimit": 2048000,
      "rowLimit": 10000
     },
     "inputWidgets": {},
     "nuid": "7a7c3cda-4418-422d-bd4f-9dcb9cd20e21",
     "showTitle": false,
     "title": ""
    }
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------------------------------------+---------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+----------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+\n|File Name                                |DOI_urls                                     |Abstract                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |DOI                         |Image                                                                       |processed_image                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |\n+-----------------------------------------+---------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+----------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+\n|ScienceDirect_citations_1715541681970.txt|https://doi.org/10.1016/j.molcel.2011.07.023.|In Escherichia coli, RNA degradation often begins with conversion of the 5′-terminal triphosphate to a monophosphate, creating a better substrate for internal cleavage by RNase E. Remarkably, no homolog of this key endonuclease is present in many bacterial species, such as Bacillus subtilis and various pathogens. Here, we report that the degradation of primary transcripts in B. subtilis can nevertheless be triggered by an analogous process to generate a short-lived, monophosphorylated intermediate. Like its E. coli counterpart, the B. subtilis RNA pyrophosphohydrolase that catalyzes this event is a Nudix protein that prefers unpaired 5′ ends. However, in B. subtilis, this modification exposes transcripts to rapid 5′ exonucleolytic degradation by RNase J, which is absent in E. coli but present in most bacteria lacking RNase E. This pathway, which closely resembles the mechanism by which deadenylated mRNA is degraded in eukaryotic cells, explains the stabilizing influence of 5′-terminal stem-loops in such bacteria.                                                                                                                                                             |10.1016/j.molcel.2011.07.023|s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.molcel.2011.07.023.jpg|[{ppp, [97, 121, 179, 163]}, {BsRppH, [512, 250, 680, 300]}, {p, [147, 411, 185, 449]}, {RNase J, [515, 545, 677, 587]}]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |\n|ScienceDirect_citations_1715541681970.txt|https://doi.org/10.1016/j.molcel.2011.07.033.|The Drosophila ecdysone receptor (EcR/Usp) is thought to activate or repress gene transcription depending on the presence or absence, respectively, of the hormone ecdysone. Unexpectedly, we found an alternative mechanism at work in salivary glands during the ecdysone-dependent transition from larvae to pupae. In the absense of ecdysone, both ecdysone receptor subunits localize to the cytoplasm, and the heme-binding nuclear receptor E75A replaces EcR/Usp at common target sequences in several genes. During the larval-pupal transition, a switch from gene activation by EcR/Usp to gene repression by E75A is triggered by a decrease in ecdysone concentration and by direct repression of the EcR gene by E75A. Additional control is provided by developmentally timed modulation of E75A activity by NO, which inhibits recruitment of the corepressor SMRTER. These results suggest a mechanism for sequential modulation of gene expression during development by competing nuclear receptors and their effector molecules, ecdysone and NO.                                                                                                                                                           |10.1016/j.molcel.2011.07.033|s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.molcel.2011.07.033.jpg|[{Ecdysone, [165, 87, 477, 173]}, {NO, [933, 93, 1031, 155]}, {Larval stages:, [0, 363, 469, 449]}, {Ligands:, [0, 463, 299, 555]}, {+20HE +NO, [329, 477, 643, 537]}, {ZOHE +NO, [760, 475, 1062, 540]}, {~ZOHE -NO, [1204, 477, 1491, 537]}, {BR-C, [829, 613, 1003, 679]}, {BR-C, [1233, 615, 1405, 679]}, {EcR, [473, 865, 611, 931]}, {E7SA, [823, 865, 993, 931]}, {NO, [1207, 855, 1309, 919]}, {0, [819, 927, 959, 1079]}, {TRR, [819, 1127, 893, 1165]}, {TRR, [1385, 1131, 1457, 1167]}, {EcR, [749, 1191, 815, 1229]}, {SMR, [909, 1193, 987, 1231]}, {EcR, [1315, 1195, 1379, 1231]}, {NO (heme), [1006, 1236, 1092, 1296]}, {TRR QOHE EcR, [329, 1271, 455, 1377]}, {SMR heme) EZSA, [1401, 1257, 1496, 1375]}, {EZSA, [905, 1331, 993, 1371]}, {Activation, [309, 1435, 623, 1497]}, {Transition, [811, 1434, 1121, 1496]}, {Repression, [1278, 1436, 1630, 1505]}, {(, [482, 921, 584, 1086]}, {Usp, [701, 1268, 763, 1322]}, {Usp, [1264, 1270, 1328, 1324]}, {Usp, [301, 1328, 364, 1380]}]                                                                               |\n|ScienceDirect_citations_1715541681970.txt|https://doi.org/10.1016/j.ccr.2011.11.003.   |Tumor immune surveillance and cancer immunotherapies are thought to depend on the intratumoral infiltration of activated CD8+ T cells. Intratumoral CD8+ T cells are rare and lack activity. IL-10 is thought to contribute to the underlying immune suppressive microenvironment. Defying those expectations we demonstrate that IL-10 induces several essential mechanisms for effective antitumor immune surveillance: infiltration and activation of intratumoral tumor-specific cytotoxic CD8+ T cells, expression of the Th1 cytokine interferon-γ (IFNγ) and granzymes in CD8+ T cells, and intratumoral antigen presentation molecules. Consequently, tumor immune surveillance is weakened in mice deficient for IL-10 whereas transgenic overexpression of IL-10 protects mice from carcinogenesis. Treatment with pegylated IL-10 restores tumor-specific intratumoral CD8+ T cell function and controls tumor growth.                                                                                                                                                                                                                                                                                                |10.1016/j.ccr.2011.11.003   |s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.ccr.2011.11.003.jpg   |[{IL-1O, [1080, 0, 1377, 107]}, {IL-10, [587, 159, 811, 247]}, {APC, [391, 337, 587, 425]}, {CD8+, [841, 415, 1081, 501]}, {T cell, [847, 515, 1082, 604]}, {Granzymes, [237, 907, 732, 1020]}, {Perforin, [309, 1011, 659, 1099]}, {MHCI, [925, 1115, 1163, 1201]}, {MHC II, [923, 1205, 1103, 1265]}, {APC, [1120, 1396, 1313, 1485]}, {Tumor, [678, 1532, 963, 1619]}, {cell, [731, 1630, 901, 1722]}, {1, [84, 476, 154, 1142]}, {IFNY, [1151, 891, 1328, 1024]}]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |\n|ScienceDirect_citations_1715541681970.txt|https://doi.org/10.1016/j.molcel.2011.06.013.|MicroRNAs (miRNAs) regulate gene expression through translation repression and mRNA destabilization. However, the molecular mechanisms of miRNA silencing are still not well defined. Using a genetic screen in mouse embryonic stem (ES) cells, we identify mammalian hyperplastic discs protein EDD, a known E3 ubiquitin ligase, as a key component of the miRNA silencing pathway. ES cells deficient for EDD are defective in miRNA function and exhibit growth defects. We demonstrate that E3 ubiquitin ligase activity is dispensable for EDD function in miRNA silencing. Instead, EDD interacts with GW182 family proteins in the Argonaute-miRNA complexes. The PABC domain of EDD is essential for its silencing function. Through the PABC domain, EDD participates in miRNA silencing by recruiting downstream effectors. Among the PABC-interactors, DDX6 and Tob1/2 are both required and sufficient for silencing mRNA targets. Taken together, these data demonstrate a critical function for EDD in miRNA silencing.                                                                                                                                                                                          |10.1016/j.molcel.2011.06.013|s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.molcel.2011.06.013.jpg|[{I, [384, 432, 458, 596]}, {PABC- Interaction network, [312, 717, 593, 902]}, {PABP, [1171, 745, 1305, 793]}, {EDD, [993, 945, 1193, 1033]}, {AAAAAAAAAAA, [271, 1395, 637, 1437]}, {AUG, [530, 189, 624, 225]}, {m7G, [220, 218, 346, 256]}, {8, [1473, 502, 1569, 558]}, {1, [1429, 716, 1514, 867]}, {1, [1346, 758, 1455, 899]}, {3, [1615, 898, 1607, 1085]}, {1, [1261, 984, 1206, 1161]}, {1, [1339, 1029, 1293, 1209]}]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |\n|ScienceDirect_citations_1715541681970.txt|https://doi.org/10.1016/j.molcel.2011.05.036.|The mechanisms by which ubiquitin ligases are regulated remain poorly understood. Here we describe a series of molecular events that coordinately regulate CHIP, a neuroprotective E3 implicated in protein quality control. Through their opposing activities, the initiator E2, Ube2w, and the specialized deubiquitinating enzyme (DUB), ataxin-3, participate in initiating, regulating, and terminating the CHIP ubiquitination cycle. Monoubiquitination of CHIP by Ube2w stabilizes the interaction between CHIP and ataxin-3, which through its DUB activity limits the length of chains attached to CHIP substrates. Upon completion of substrate ubiquitination, ataxin-3 deubiquitinates CHIP, effectively terminating the reaction. Our results suggest that functional pairing of E3s with ataxin-3 or similar DUBs represents an important point of regulation in ubiquitin-dependent protein quality control. In addition, the results shed light on disease pathogenesis in SCA3, a neurodegenerative disorder caused by polyglutamine expansion in ataxin-3.                                                                                                                                                    |10.1016/j.molcel.2011.05.036|s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.molcel.2011.05.036.jpg|[{Ubezw ubiquitinates CHIP stabilizing the CHIPlataxin-3 interaction, [0, 0, 567, 71]}, {Ataxin-3 trims polyubiquitin chains on CHIP substrates and deubiquitinates CHIP in response to substrate ubiquitination, [705, 0, 1639, 71]}, {Ubezw, [364, 116, 468, 148]}, {chaln extending 6z, [1113, 152, 1214, 224]}, {HSP, [112, 314, 176, 346]}, {HSP, [1240, 314, 1304, 344]}, {Ubezw, [608, 337, 713, 376]}, {loeed pictln, [64, 391, 187, 465]}, {maaued polnt, [1180, 392, 1318, 463]}, {C#, [1293, 162, 1380, 151]}, {0, [175, 167, 255, 155]}, {ctot, [419, 318, 474, 419]}]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |\n|ScienceDirect_citations_1715541681970.txt|https://doi.org/10.1016/j.cmet.2011.08.015.  |Mammalian cells express dozens of iron-containing proteins, yet little is known about the mechanism of metal ligand incorporation. Human poly (rC) binding protein 1 (PCBP1) is an iron chaperone that binds iron and delivers it to ferritin, a cytosolic iron storage protein. We have identified the iron-dependent prolyl hydroxylases (PHDs) and asparaginyl hydroxylase (FIH1) that modify hypoxia-inducible factor α (HIFα) as targets of PCBP1. Depletion of PCBP1 or PCBP2 in cells led to loss of PHD activity, manifested by reduced prolyl hydroxylation of HIF1α, impaired degradation of HIF1α through the VHL/proteasome pathway, and accumulation of active HIF1 transcription factor. PHD activity was restored in vitro by addition of excess Fe(II), or purified Fe-PCBP1, and PCBP1 bound to PHD2 and FIH1 in vivo. These data indicated that PCBP1 was required for iron incorporation into PHD and suggest a broad role for PCBP1 and 2 in delivering iron to cytosolic nonheme iron enzymes.                                                                                                                                                                                                              |10.1016/j.cmet.2011.08.015  |s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.cmet.2011.08.015.jpg  |[{PHDs FIHT, [1192, 157, 1341, 278]}, {2-OG Fe Dioxygenases other Fe enzymes?, [966, 344, 1518, 480]}, {PCBP182, [54, 607, 295, 663]}, {Fe(ll), [336, 608, 475, 673]}, {Apo-Ferritin, [875, 851, 1157, 917]}, {Holo-Ferritin, [1272, 850, 1566, 906]}]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |\n|ScienceDirect_citations_1715541681970.txt|https://doi.org/10.1016/j.molcel.2011.06.038.|The signaling adaptor p62 is a critical mediator of important cellular functions, owing to its ability to establish interactions with various signaling intermediaries. Here, we identify raptor as an interacting partner of p62. Thus, p62 is an integral part of the mTORC1 complex and is necessary to mediate amino acid signaling for the activation of S6K1 and 4EBP1. p62 interacts in an amino acid-dependent manner with mTOR and raptor. In addition, p62 binds the Rags proteins and favors formation of the active Rag heterodimer that is further stabilized by raptor. Interestingly, p62 colocalizes with Rags at the lysosomal compartment and is required for the interaction of mTOR with Rag GTPases in vivo and for translocation of the mTORC1 complex to the lysosome, a crucial step for mTOR activation.                                                                                                                                                                                                                                                                                                                                                                                                |10.1016/j.molcel.2011.06.038|s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.molcel.2011.06.038.jpg|[{raptor, [280, 266, 494, 346]}, {raptor, [1354, 421, 1576, 514]}, {1, [1495, 559, 1579, 845]}, {amino, [717, 867, 935, 935]}, {lysosome, [130, 930, 472, 1008]}, {lysosome, [1142, 938, 1482, 1016]}, {acids, [712, 1012, 906, 1084]}, {mTORC1 (inactive), [189, 1092, 514, 1239]}, {mTORCI (active), [1208, 1102, 1512, 1245]}, {8, [296, 22, 210, 255]}, {8, [1365, 205, 1280, 439]}, {1, [467, 564, 387, 863]}, {1, [1415, 576, 1355, 883]}, {L, [624, 588, 630, 901]}, {p62, [109, 626, 245, 691]}, {p62, [1130, 642, 1268, 704]}]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |\n|ScienceDirect_citations_1715541681970.txt|https://doi.org/10.1016/j.molcel.2011.08.037.|Histone acetyltransferase (HAT) complexes are coactivators that are important for transcriptional activation by modifying chromatin. Metazoan SAGA and ATAC are distinct multisubunits complexes that share the same catalytic HAT subunit (GCN5 or PCAF). Here, we show that these human HAT complexes are targeted to different genomic loci representing functionally distinct regulatory elements both at broadly expressed and tissue-specific genes. While SAGA can principally be found at promoters, ATAC is recruited to promoters and enhancers, yet only its enhancer binding is\n\n*** WARNING: max output size exceeded, skipping output. ***\n\n{Kinase Active, [160, 860, 666, 936]}, {Synaptic function, [991, 1282, 1642, 1379]}, {Memory formation, [993, 1403, 1660, 1503]}]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |\n|ScienceDirect_citations_1715541681970.txt|https://doi.org/10.1016/j.devcel.2011.09.001.|Faithful segregation of homologous chromosomes during meiosis requires pairing, synapsis, and crossing-over. In C. elegans, homolog pairing and synapsis depend on pairing centers (PCs), special regions near one end of each chromosome that interact with the nuclear envelope (NE) and cytoplasmic microtubules. Here, we report that PCs are required for nuclear reorganization at the onset of meiosis. We demonstrate that PCs recruit the Polo-like kinase PLK-2 to induce NE remodeling, chromosome pairing, and synapsis. Recruitment of PLK-2 is also required to mediate a cell cycle delay and selective apoptosis of nuclei containing unsynapsed chromosomes, establishing a molecular link between these two quality control mechanisms. This work reveals unexpected functions for the conserved family of Polo-like kinases, and advances our understanding of how meiotic processes are properly coordinated to ensure transmission of genetic information from parents to progeny.                                                                                                                                                                                                                          |10.1016/j.devcel.2011.09.001|s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.devcel.2011.09.001.jpg|[{Melotic entry: Chromosome palrlng centers recrult the Polo-Ilke klnase PLK-2, [158, 1, 1555, 131]}, {dynein, [1405, 165, 1550, 216]}, {PLK-2, [479, 407, 609, 451]}, {HIM-BIZIM, [455, 487, 659, 531]}, {chromosome pairing center, [657, 537, 951, 643]}, {Phosphorylation of the nuclear envelope proteln SUN-1 promotes SUN-1IZYG-12 aggregatlon, homolog palring; and synapsls, [102, 779, 1551, 913]}, {microtubule, [1355, 973, 1607, 1033]}, {pSer12, [1305, 1267, 1465, 1336]}, {ZYG-12, [302, 164, 464, 196]}, {ONM, [72, 234, 222, 243]}, {SUN-1, [319, 293, 461, 323]}, {INM, [156, 433, 277, 465]}]                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |\n|ScienceDirect_citations_1715541681970.txt|https://doi.org/10.1016/j.devcel.2011.10.009.|Vesicle transport requires four steps: vesicle formation, movement, tethering, and fusion. In yeast, two Rab GTPases, Ypt31/32, are required for post-Golgi vesicle formation. A third Rab GTPase, Sec4, and the exocyst act in tethering and fusion of these vesicles. Vesicle production is coupled to transport via direct interaction between Ypt31/32 and the yeast myosin V, Myo2. Here we show that Myo2 interacts directly with Sec4 and the exocyst subunit Sec15. Disruption of these interactions results in compromised growth and the accumulation of secretory vesicles. We identified the Sec15-binding region on Myo2 and also identified residues on Sec15 required for interaction with Myo2. That Myo2 interacts with Sec15 uncovers additional roles for the exocyst as an adaptor for molecular motors and implies similar roles for structurally related tethering complexes. Moreover, these studies predict that for many pathways, molecular motors attach to vesicles prior to their formation and remain attached until fusion.                                                                                                                                                                       |10.1016/j.devcel.2011.10.009|s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.devcel.2011.10.009.jpg|[{actin, [393, 55, 489, 95]}, {Sc Sec6, [1466, 118, 1660, 170]}, {Myo2, [410, 166, 546, 222]}, {2) movement, [377, 234, 684, 295]}, {3) tethering, [1012, 216, 1280, 291]}, {1) vesicle production, [0, 312, 245, 429]}, {Dm Sec15, [1422, 380, 1660, 430]}, {Ypt31/32, [279, 421, 452, 478]}, {Sec4, [529, 413, 636, 463]}, {exocyst, [784, 480, 990, 545]}, {Sc Exo70, [1440, 500, 1660, 556]}, {trans-Golgi, [136, 620, 352, 674]}, {4) fusion, [814, 690, 1016, 744]}, {Mm Exo70, [1416, 678, 1660, 734]}, {Myoz-CBD, [468, 759, 725, 825]}, {plasma membrane, [909, 837, 1117, 939]}, {Sc Exo84, [1440, 923, 1660, 980]}]                                                                                                                                                                                                                                                                                                                                                                                                                                                             |\n|ScienceDirect_citations_1715541681970.txt|https://doi.org/10.1016/j.devcel.2011.07.020.|The netrins and slits are two families of widely conserved cues that guide axons and cells along the dorsal-ventral (D-V) axis of animals. These cues typically emanate from the dorsal or ventral midlines and provide spatial information to migrating cells by forming gradients along the D-V axis. Some cell types, however, extend processes to both the dorsal and ventral midlines, suggesting the existence of additional guidance cues that are secreted from both midlines. Here, we report that a previously uncharacterized protein called MADD-4 is secreted by the dorsal and ventral nerve cords of the nematode C. elegans to attract sensory axons and muscle membrane extensions called muscle arms. MADD-4's activity is dependent on UNC-40/DCC, a netrin receptor, which functions cell-autonomously to direct membrane extension. The biological role of MADD-4 orthologs, including ADAMTSL1 and 3 in mammals, is unknown. MADD-4 may therefore represent the founding member of a family of guidance proteins.                                                                                                                                                                                          |10.1016/j.devcel.2011.07.020|s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.devcel.2011.07.020.jpg|[{midline motor axons, [13, 35, 817, 123]}, {UNC-40, [657, 247, 963, 333]}, {(DCC), [683, 325, 919, 425]}, {UNC-6, [388, 460, 644, 538]}, {MADD-4, [915, 463, 1256, 544]}, {(netrin), [360, 530, 668, 630]}, {actin polymerization &, [351, 941, 1247, 1055]}, {extension to the midline, [323, 1041, 1281, 1129]}, {1, [1556, 926, 1636, 1278]}, {muscle arm, [331, 1211, 784, 1292]}]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |\n|ScienceDirect_citations_1715541681970.txt|https://doi.org/10.1016/j.devcel.2011.09.019.|Homologous chromosome pairing is a prerequisite to establish physical linkage between homologs, which is critical for faithful chromosome segregation during meiosis I. The establishment of pairing is genetically separable from subsequent synapsis, defined as stabilization of pairing by the synaptonemal complex (SC). The underlying mechanism of presynaptic pairing is poorly understood. In the nematode Caenorhabditis elegans, a unique cis-acting element, the pairing center (PC), is essential for presynaptic pairing; however, it is not known whether and how the remainder of the chromosome contributes to presynaptic pairing. Here we report direct evidence for presynaptic pairing activity intrinsic to non-PC regions, which is facilitated by a conserved chromodomain protein, MRG-1. In mrg-1 loss-of-function mutants, pairing is compromised specifically in non-PC regions, leading to nonhomologous SC assembly. Our data support a model in which presynaptic alignment in non-PC regions collaborates with initial PC pairing to ensure correct homologous synapsis.                                                                                                                         |10.1016/j.devcel.2011.09.019|s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.devcel.2011.09.019.jpg|[{PC, [10, 70, 154, 166]}, {MRG-1, [516, 72, 836, 168]}, {Initial, [6, 168, 210, 248]}, {pairing, [6, 264, 270, 364]}, {Alignment, [447, 493, 836, 597]}, {Synapsis, [962, 492, 1315, 601]}, {SC, [1486, 472, 1636, 574]}, {MRG, [529, 1059, 765, 1151]}, {Non-homologous, [278, 1367, 927, 1488]}, {synapsis, [274, 1477, 621, 1590]}]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |\n|ScienceDirect_citations_1715541681970.txt|https://doi.org/10.1016/j.cmet.2011.06.017.  |In proliferating cells, a transition from aerobic to anaerobic metabolism is known as the Warburg effect, whose reversal inhibits cancer cell proliferation. Studying its regulator pyruvate kinase (PYK) in yeast, we discovered that central metabolism is self-adapting to synchronize redox metabolism when respiration is activated. Low PYK activity activated yeast respiration. However, levels of reactive oxygen species (ROS) did not increase, and cells gained resistance to oxidants. This adaptation was attributable to accumulation of the PYK substrate phosphoenolpyruvate (PEP). PEP acted as feedback inhibitor of the glycolytic enzyme triosephosphate isomerase (TPI). TPI inhibition stimulated the pentose phosphate pathway, increased antioxidative metabolism, and prevented ROS accumulation. Thus, a metabolic feedback loop, initiated by PYK, mediated by its substrate and acting on TPI, stimulates redox metabolism in respiring cells. Originating from a single catalytic step, this autonomous reconfiguration of central carbon metabolism prevents oxidative stress upon shifts between fermentation and respiration.                                                                   |10.1016/j.cmet.2011.06.017  |s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.cmet.2011.06.017.jpg  |[{NADP+, [625, 66, 797, 140]}, {NADPH, [1006, 94, 1174, 144]}, {PPP, [572, 280, 716, 356]}, {3, [555, 513, 593, 567]}, {1, [128, 417, 238, 779]}, {Tpi, [376, 692, 452, 746]}, {1, [1224, 634, 1272, 710]}, {1, [1222, 722, 1274, 804]}, {2, [233, 843, 271, 897]}, {PEP, [449, 929, 603, 1013]}, {4, [1075, 949, 1111, 995]}, {{, [1190, 803, 1279, 1022]}, {Pyk low activity in yeast, [403, 1083, 634, 1244]}, {ROS, [996, 1084, 1242, 1206]}, {fermentation, [121, 1377, 455, 1437]}, {respiration, [588, 1447, 1114, 1620]}, {ROS, [1217, 1485, 1359, 1559]}]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |\n|ScienceDirect_citations_1715541681970.txt|https://doi.org/10.1016/j.devcel.2011.06.029.|In vertebrates, unfertilized eggs are arrested at metaphase of meiosis II by Mos and Emi2, an inhibitor of the APC/C ubiquitin ligase. In Xenopus, Cdk1 phosphorylates Emi2 and both destabilizes and inactivates it, whereas Mos recruits PP2A phosphatase to antagonize the Cdk1 phosphorylation. However, how Cdk1 phosphorylation inhibits Emi2 is largely unknown. Here we show that multiple N-terminal Cdk1 phosphorylation motifs bind cyclin B1-Cdk1 itself, Plk1, and CK1δ/ε to inhibit Emi2. Plk1, after rebinding to other sites by self-priming phosphorylation, partially destabilizes Emi2. Cdk1 and CK1δ/ε sequentially phosphorylate the C-terminal APC/C-docking site, thereby cooperatively inhibiting Emi2 from binding the APC/C. In the presence of Mos, however, PP2A-B56β/ε bind to Emi2 and keep dephosphorylating it, particularly at the APC/C-docking site. Thus, Emi2 stability and activity are dynamically regulated by Emi2-bound multiple kinases and PP2A phosphatase. Our data also suggest a general role for Cdk1 substrate phosphorylation motifs in M phase regulation.                                                                                                                   |10.1016/j.devcel.2011.06.029|s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.devcel.2011.06.029.jpg|[{Inactivation, [841, 485, 1041, 527]}, {Cdk1, [221, 605, 319, 649]}, {PPZA- Bs6ple, [876, 617, 1008, 708]}, {Plk1, [385, 661, 467, 703]}, {CK1Se, [503, 663, 625, 703]}, {Cdk1, [648, 660, 746, 702]}, {P, [340, 722, 368, 750]}, {P, [470, 720, 496, 748]}, {PJP), [1129, 719, 1205, 755]}, {N, [169, 795, 205, 835]}, {3&e:elg/a 882, [328, 768, 778, 870]}, {82, [895, 769, 981, 859]}, {28, [1127, 769, 1211, 857]}, {88, [1338, 770, 1412, 858]}, {C, [1443, 795, 1481, 837]}, {DSG, [247, 871, 327, 907]}, {M, [402, 874, 436, 906]}, {N, [678, 874, 706, 904]}, {DSA, [787, 871, 863, 907]}, {D-box, [1059, 873, 1165, 909]}, {ZBR, [1231, 871, 1311, 907]}, {RL, [1355, 871, 1407, 907]}, {Plk1, [507, 927, 587, 969]}, {Rsk, [897, 941, 973, 983]}, {Destab, [318, 965, 446, 1005]}, {Destab., [646, 965, 776, 1007]}, {SCFB-TrCP, [476, 1027, 627, 1072]}, {APCIC, [1181, 995, 1363, 1057]}, {Mos, [893, 1073, 976, 1117]}]                                                                                                                                                     |\n|ScienceDirect_citations_1715541681970.txt|https://doi.org/10.1016/j.molcel.2011.09.026.|The functional significance of the signaling pathway induced by O6-methylguanine (O6-MeG) lesions is poorly understood. Here, we identify the p50 subunit of NF-κB as a central target in the response to O6-MeG and demonstrate that p50 is required for SN1-methylator-induced cytotoxicity. In response to SN1-methylation, p50 facilitates the inhibition of NF-κB-regulated antiapoptotic gene expression. Inhibition of NF-κB activity is noted to be an S phase-specific phenomenon that requires the formation of O6-MeG:T mismatches. Chk1 associates with p50 following SN1-methylation, and phosphorylation of p50 by Chk1 results in the inhibition of NF-κB DNA binding. Expression of an unphosphorylatable p50 mutant blocks inhibition of NF-κB-regulated antiapoptotic gene expression and attenuates SN1-methylator-induced cytotoxicity. While O6-MeG:T-induced, p50-dependent signaling is not sufficient to induce cell death, this pathway sensitizes cells to the cytotoxic effects of DNA breaks.                                                                                                                                                                                                        |10.1016/j.molcel.2011.09.026|s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.molcel.2011.09.026.jpg|[{A, [937, 33, 1007, 93]}, {DNA Replication, [473, 207, 807, 267]}, {MMR Apparatus, [278, 396, 513, 509]}, {Ae, [947, 457, 1015, 517]}, {Bcl-xL Cox-2, [220, 682, 354, 792]}, {DNA Replication, [1295, 659, 1611, 721]}, {NF-KB, [10, 694, 148, 744]}, {Chk), [468, 840, 578, 890]}, {D I, [1166, 834, 1380, 906]}, {TI, [1204, 922, 1326, 996]}, {Bcl-xL Cox-2, [358, 1159, 494, 1272]}, {Cytotoxic DNA DSB Response, [1008, 1152, 1629, 1220]}, {Decreased Pro-Survival Gene Expression, [394, 1386, 884, 1510]}, {CELL DEATH, [1191, 1581, 1521, 1641]}, {~Me, [928, 249, 1016, 279]}, {VTI Iz, [1305, 942, 1588, 877]}]                                                                                                                                                                                                                                                                                                                                                                                                                                                                |\n|ScienceDirect_citations_1715541681970.txt|https://doi.org/10.1016/j.devcel.2011.07.013.|Neutrophil homeostasis is essential for host defense. Here we identify dual roles for Rac2 during neutrophil homeostasis using a zebrafish model of primary immune deficiency induced by the human inhibitory Rac2D57N mutation in neutrophils. Noninvasive live imaging of Rac2 morphants or Rac2D57N zebrafish larvae demonstrates an essential role for Rac2 in regulating 3D motility and the polarization of F-actin dynamics and PI(3)K signaling in vivo. Tracking of photolabeled Rac2-deficient neutrophils from hematopoietic tissue also shows increased mobilization into the circulation, indicating that neutrophil mobilization does not require traditionally defined cell motility. Moreover, excessive neutrophil retention in hematopoietic tissue resulting from a constitutively active CXCR4 mutation in zebrafish warts, hypogammaglobulinemia, infections, and myelokathexis (WHIM) syndrome is partially rescued by the inhibitory Rac2 mutation. These findings reveal that Rac2 signaling is necessary for both neutrophil 3D motility and CXCR4-mediated neutrophil retention in hematopoietic tissue, thereby limiting neutrophil mobilization, a critical first step in the innate immune response.|10.1016/j.devcel.2011.07.013|s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.devcel.2011.07.013.jpg|[{Or, [259, 161, 317, 205]}, {8, [0, 555, 54, 671]}, {Hematopoietic tissue, [253, 691, 737, 753]}, {site of infection, [1138, 689, 1491, 746]}, {Rac2 dependent retention, [94, 790, 832, 866]}, {Rac2 dependent infiltration, [888, 789, 1660, 864]}, {Neutrophil, [165, 953, 413, 1017]}, {Macrophage#CXCRA, [519, 947, 1038, 1021]}]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |\n|ScienceDirect_citations_1715541681970.txt|https://doi.org/10.1016/j.ccr.2011.08.023.   |A genomic signature designed to assess the activity of the estrogen-related receptor alpha (ERRα) was used to profile more than 800 breast tumors, revealing a shorter disease-free survival in patients with tumors exhibiting elevated receptor activity. Importantly, this signature also predicted the ability of an ERRα antagonist, XCT790, to inhibit proliferation in cellular models of breast cancer. Using a chemical genomic approach, it was determined that activation of the Her2/IGF-1R signaling pathways and subsequent C-MYC stabilization upregulate the expression of peroxisome proliferator-activated receptor gamma coactivator-1 beta (PGC-1β), an obligate cofactor for ERRα activity. PGC-1β knockdown in breast cancer cells impaired ERRα signaling and reduced cell proliferation, implicating a functional role for PGC-1β/ERRα in the pathogenesis of breast cancers.                                                                                                                                                                                                                                                                                                                            |10.1016/j.ccr.2011.08.023   |s3a://scisketch-dataset/Graphical-Abstracts/10.1016xj.ccr.2011.08.023.jpg   |[{The PI3K Pathway is Upstream of PGC-1BIERRa, [87, 0, 1579, 85]}, {HER2, [273, 245, 404, 298]}, {IGFI-R, [1286, 244, 1444, 292]}, {Pisk, [794, 278, 900, 328]}, {L, [958, 262, 1129, 329]}, {AKT, [810, 440, 904, 488]}, {cmye, [63, 631, 191, 675]}, {PGC-1p, [215, 722, 392, 785]}, {PGC-1p}, [465, 718, 640, 785]}, {ERRA, [926, 710, 1058, 760]}, {PGC-1R, [701, 848, 874, 910]}, {ERR\", [734, 946, 866, 996]}, {Transcriptan, [900, 1120, 1160, 1176]}]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |\n+-----------------------------------------+---------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+----------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+\nonly showing top 20 rows\n\n"
     ]
    }
   ],
   "source": [
    "import boto3\n",
    "import easyocr\n",
    "import cv2\n",
    "import numpy as np\n",
    "from sklearn.cluster import DBSCAN\n",
    "import torch\n",
    "import io\n",
    "from PIL import Image\n",
    "import pandas as pd\n",
    "from pyspark.sql import SparkSession\n",
    "from pyspark.sql.functions import udf\n",
    "from pyspark.sql.types import StringType, ArrayType, StructType, StructField, IntegerType\n",
    "\n",
    "\n",
    "# Function to process images directly from S3 (used in the UDF)\n",
    "def process_image_from_s3(image_path):\n",
    "    s3_client = boto3.client('s3')  # Initialize the S3 client inside the UDF\n",
    "    \n",
    "    s3_prefix = 's3a://scisketch-dataset/Graphical-Abstracts/'\n",
    "    if not image_path.startswith(s3_prefix):\n",
    "        raise ValueError(f\"Invalid S3 path for the given bucket and folder: {image_path}\")\n",
    "\n",
    "    image_key = image_path[len(s3_prefix):]\n",
    "\n",
    "    # Download the image from S3\n",
    "    image_object = s3_client.get_object(Bucket='scisketch-dataset', Key='Graphical-Abstracts/' + image_key)\n",
    "    image_content = image_object['Body'].read()\n",
    "\n",
    "    # Convert the image content to a numpy array\n",
    "    image = np.array(Image.open(io.BytesIO(image_content)))\n",
    "\n",
    "    # Check if the image was loaded successfully\n",
    "    if image is None:\n",
    "        print(f\"Error: Could not load image from path {image_path}\")\n",
    "        return None\n",
    "\n",
    "    original_image = image.copy()\n",
    "\n",
    "    # Initialize the EasyOCR reader with GPU support if available\n",
    "    if torch.backends.mps.is_available():\n",
    "        reader = easyocr.Reader(['en'], gpu=True)  # Use MPS (Metal Performance Shaders) backend\n",
    "    else:\n",
    "        reader = easyocr.Reader(['en'], gpu=False)\n",
    "\n",
    "    # Perform OCR on the image\n",
    "    results = reader.readtext(original_image)\n",
    "\n",
    "    # Extract the bounding box coordinates and text\n",
    "    boxes = [item[0] for item in results]\n",
    "    texts = [item[1] for item in results]\n",
    "\n",
    "    # Draw bounding boxes and texts on the image\n",
    "    for box, text in zip(boxes, texts):\n",
    "        # Draw the bounding box\n",
    "        top_left = tuple(map(int, box[0]))\n",
    "        bottom_right = tuple(map(int, box[2]))\n",
    "        cv2.rectangle(original_image, top_left, bottom_right, (0, 255, 0), 2)\n",
    "\n",
    "        # Put the OCR text near the bounding box\n",
    "        cv2.putText(original_image, text, (top_left[0], top_left[1] - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 2)\n",
    "\n",
    "    # Calculate the center of each bounding box\n",
    "    centers = np.array([[(box[0][0] + box[2][0]) / 2, (box[0][1] + box[2][1]) / 2] for box in boxes])\n",
    "\n",
    "     # Check if centers is empty\n",
    "    if centers.size == 0:\n",
    "        return None\n",
    "    \n",
    "    # Cluster the bounding boxes using DBSCAN\n",
    "    clustering = DBSCAN(eps=70, min_samples=1).fit(centers)\n",
    "\n",
    "    # Group the boxes and texts by cluster\n",
    "    clusters = {}\n",
    "    for idx, label in enumerate(clustering.labels_):\n",
    "        if label not in clusters:\n",
    "            clusters[label] = {\n",
    "                'boxes': [],\n",
    "                'texts': []\n",
    "            }\n",
    "        clusters[label]['boxes'].append(boxes[idx])\n",
    "        clusters[label]['texts'].append(texts[idx])\n",
    "\n",
    "    text_annotations = []\n",
    "    for cluster in clusters.values():\n",
    "        combined_text = ' '.join(cluster['texts'])\n",
    "\n",
    "        # Get the bounding box of the cluster\n",
    "        x_min = min([box[0][0] for box in cluster['boxes']])\n",
    "        y_min = min([box[0][1] for box in cluster['boxes']])\n",
    "        x_max = max([box[2][0] for box in cluster['boxes']])\n",
    "        y_max = max([box[2][1] for box in cluster['boxes']])\n",
    "\n",
    "        # Convert to integers\n",
    "        x_min, y_min, x_max, y_max = map(int, [x_min, y_min, x_max, y_max])\n",
    "\n",
    "        # Append to the list of text annotations\n",
    "        text_annotations.append({\n",
    "            'text': combined_text,\n",
    "            'coordinates': [x_min, y_min, x_max, y_max]\n",
    "        })\n",
    "\n",
    "    return text_annotations\n",
    "\n",
    "# Register UDF\n",
    "process_image_udf = udf(process_image_from_s3, ArrayType(StructType([\n",
    "    StructField(\"text\", StringType(), True),\n",
    "    StructField(\"coordinates\", ArrayType(IntegerType()), True)\n",
    "])))\n",
    "\n",
    "# Apply the UDF to process images\n",
    "processed_spark_df = spark_df.withColumn(\"processed_image\", process_image_udf(spark_df.Image))\n",
    "\n",
    "# Show the results\n",
    "processed_spark_df.show(truncate=False)\n",
    "\n",
    "# Convert to JSON format and save the output to S3 with overwrite mode\n",
    "output_path = \"s3a://scisketch-dataset/processed_data/processed_data.json\"\n",
    "processed_spark_df.write.mode(\"overwrite\").json(output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "metadata": {
    "application/vnd.databricks.v1+cell": {
     "cellMetadata": {
      "byteLimit": 2048000,
      "rowLimit": 10000
     },
     "inputWidgets": {},
     "nuid": "38118a03-e823-41e1-aaf2-700e24686a45",
     "showTitle": false,
     "title": ""
    }
   },
   "outputs": [
    {
     "output_type": "display_data",
     "data": {
      "text/plain": [
       "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m\n",
       "\u001B[0;31mJSONDecodeError\u001B[0m                           Traceback (most recent call last)\n",
       "File \u001B[0;32m<command-3998437280854561>, line 26\u001B[0m\n",
       "\u001B[1;32m     24\u001B[0m     json_content \u001B[38;5;241m=\u001B[39m obj[\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mBody\u001B[39m\u001B[38;5;124m'\u001B[39m]\u001B[38;5;241m.\u001B[39mread()\u001B[38;5;241m.\u001B[39mdecode(\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mutf-8\u001B[39m\u001B[38;5;124m'\u001B[39m)\n",
       "\u001B[1;32m     25\u001B[0m     \u001B[38;5;66;03m# Parse the JSON content and append to the list\u001B[39;00m\n",
       "\u001B[0;32m---> 26\u001B[0m     json_objects\u001B[38;5;241m.\u001B[39mextend(json\u001B[38;5;241m.\u001B[39mloads(json_content))\n",
       "\u001B[1;32m     28\u001B[0m \u001B[38;5;66;03m# Convert the list of JSON objects to a pandas DataFrame\u001B[39;00m\n",
       "\u001B[1;32m     29\u001B[0m df \u001B[38;5;241m=\u001B[39m pd\u001B[38;5;241m.\u001B[39mDataFrame(json_objects)\n",
       "\n",
       "File \u001B[0;32m/usr/lib/python3.10/json/__init__.py:346\u001B[0m, in \u001B[0;36mloads\u001B[0;34m(s, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)\u001B[0m\n",
       "\u001B[1;32m    341\u001B[0m     s \u001B[38;5;241m=\u001B[39m s\u001B[38;5;241m.\u001B[39mdecode(detect_encoding(s), \u001B[38;5;124m'\u001B[39m\u001B[38;5;124msurrogatepass\u001B[39m\u001B[38;5;124m'\u001B[39m)\n",
       "\u001B[1;32m    343\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m (\u001B[38;5;28mcls\u001B[39m \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m \u001B[38;5;129;01mand\u001B[39;00m object_hook \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m \u001B[38;5;129;01mand\u001B[39;00m\n",
       "\u001B[1;32m    344\u001B[0m         parse_int \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m \u001B[38;5;129;01mand\u001B[39;00m parse_float \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m \u001B[38;5;129;01mand\u001B[39;00m\n",
       "\u001B[1;32m    345\u001B[0m         parse_constant \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m \u001B[38;5;129;01mand\u001B[39;00m object_pairs_hook \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m \u001B[38;5;129;01mand\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m kw):\n",
       "\u001B[0;32m--> 346\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43m_default_decoder\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mdecode\u001B[49m\u001B[43m(\u001B[49m\u001B[43ms\u001B[49m\u001B[43m)\u001B[49m\n",
       "\u001B[1;32m    347\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;28mcls\u001B[39m \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m:\n",
       "\u001B[1;32m    348\u001B[0m     \u001B[38;5;28mcls\u001B[39m \u001B[38;5;241m=\u001B[39m JSONDecoder\n",
       "\n",
       "File \u001B[0;32m/usr/lib/python3.10/json/decoder.py:340\u001B[0m, in \u001B[0;36mJSONDecoder.decode\u001B[0;34m(self, s, _w)\u001B[0m\n",
       "\u001B[1;32m    338\u001B[0m end \u001B[38;5;241m=\u001B[39m _w(s, end)\u001B[38;5;241m.\u001B[39mend()\n",
       "\u001B[1;32m    339\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m end \u001B[38;5;241m!=\u001B[39m \u001B[38;5;28mlen\u001B[39m(s):\n",
       "\u001B[0;32m--> 340\u001B[0m     \u001B[38;5;28;01mraise\u001B[39;00m JSONDecodeError(\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mExtra data\u001B[39m\u001B[38;5;124m\"\u001B[39m, s, end)\n",
       "\u001B[1;32m    341\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m obj\n",
       "\n",
       "\u001B[0;31mJSONDecodeError\u001B[0m: Extra data: line 2 column 1 (char 1495)"
      ]
     },
     "metadata": {
      "application/vnd.databricks.v1+output": {
       "arguments": {},
       "data": "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m\n\u001B[0;31mJSONDecodeError\u001B[0m                           Traceback (most recent call last)\nFile \u001B[0;32m<command-3998437280854561>, line 26\u001B[0m\n\u001B[1;32m     24\u001B[0m     json_content \u001B[38;5;241m=\u001B[39m obj[\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mBody\u001B[39m\u001B[38;5;124m'\u001B[39m]\u001B[38;5;241m.\u001B[39mread()\u001B[38;5;241m.\u001B[39mdecode(\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mutf-8\u001B[39m\u001B[38;5;124m'\u001B[39m)\n\u001B[1;32m     25\u001B[0m     \u001B[38;5;66;03m# Parse the JSON content and append to the list\u001B[39;00m\n\u001B[0;32m---> 26\u001B[0m     json_objects\u001B[38;5;241m.\u001B[39mextend(json\u001B[38;5;241m.\u001B[39mloads(json_content))\n\u001B[1;32m     28\u001B[0m \u001B[38;5;66;03m# Convert the list of JSON objects to a pandas DataFrame\u001B[39;00m\n\u001B[1;32m     29\u001B[0m df \u001B[38;5;241m=\u001B[39m pd\u001B[38;5;241m.\u001B[39mDataFrame(json_objects)\n\nFile \u001B[0;32m/usr/lib/python3.10/json/__init__.py:346\u001B[0m, in \u001B[0;36mloads\u001B[0;34m(s, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)\u001B[0m\n\u001B[1;32m    341\u001B[0m     s \u001B[38;5;241m=\u001B[39m s\u001B[38;5;241m.\u001B[39mdecode(detect_encoding(s), \u001B[38;5;124m'\u001B[39m\u001B[38;5;124msurrogatepass\u001B[39m\u001B[38;5;124m'\u001B[39m)\n\u001B[1;32m    343\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m (\u001B[38;5;28mcls\u001B[39m \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m \u001B[38;5;129;01mand\u001B[39;00m object_hook \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m \u001B[38;5;129;01mand\u001B[39;00m\n\u001B[1;32m    344\u001B[0m         parse_int \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m \u001B[38;5;129;01mand\u001B[39;00m parse_float \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m \u001B[38;5;129;01mand\u001B[39;00m\n\u001B[1;32m    345\u001B[0m         parse_constant \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m \u001B[38;5;129;01mand\u001B[39;00m object_pairs_hook \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m \u001B[38;5;129;01mand\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m kw):\n\u001B[0;32m--> 346\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43m_default_decoder\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mdecode\u001B[49m\u001B[43m(\u001B[49m\u001B[43ms\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m    347\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;28mcls\u001B[39m \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m:\n\u001B[1;32m    348\u001B[0m     \u001B[38;5;28mcls\u001B[39m \u001B[38;5;241m=\u001B[39m JSONDecoder\n\nFile \u001B[0;32m/usr/lib/python3.10/json/decoder.py:340\u001B[0m, in \u001B[0;36mJSONDecoder.decode\u001B[0;34m(self, s, _w)\u001B[0m\n\u001B[1;32m    338\u001B[0m end \u001B[38;5;241m=\u001B[39m _w(s, end)\u001B[38;5;241m.\u001B[39mend()\n\u001B[1;32m    339\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m end \u001B[38;5;241m!=\u001B[39m \u001B[38;5;28mlen\u001B[39m(s):\n\u001B[0;32m--> 340\u001B[0m     \u001B[38;5;28;01mraise\u001B[39;00m JSONDecodeError(\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mExtra data\u001B[39m\u001B[38;5;124m\"\u001B[39m, s, end)\n\u001B[1;32m    341\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m obj\n\n\u001B[0;31mJSONDecodeError\u001B[0m: Extra data: line 2 column 1 (char 1495)",
       "errorSummary": "<span class='ansi-red-fg'>JSONDecodeError</span>: Extra data: line 2 column 1 (char 1495)",
       "errorTraceType": "ansi",
       "metadata": {},
       "type": "ipynbError"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import boto3\n",
    "import pandas as pd\n",
    "import json\n",
    "\n",
    "# Initialize the S3 client\n",
    "s3_client = boto3.client('s3')\n",
    "\n",
    "# Define the S3 bucket and prefix for the JSON files\n",
    "bucket_name = 'scisketch-dataset'\n",
    "prefix = 'processed_data/'\n",
    "\n",
    "# List all the parts in the S3 bucket\n",
    "response = s3_client.list_objects_v2(Bucket=bucket_name, Prefix=prefix)\n",
    "\n",
    "# Extract the list of JSON file keys\n",
    "json_keys = [content['Key'] for content in response.get('Contents', []) if content['Key'].endswith('.json')]\n",
    "\n",
    "# Initialize an empty list to hold the JSON objects\n",
    "json_objects = []\n",
    "\n",
    "# Read each JSON file from S3 and append the content to the list\n",
    "for key in json_keys:\n",
    "    obj = s3_client.get_object(Bucket=bucket_name, Key=key)\n",
    "    json_content = obj['Body'].read().decode('utf-8')\n",
    "    # Parse the JSON content and append to the list\n",
    "    json_objects.extend(json.loads(json_content))\n",
    "\n",
    "# Convert the list of JSON objects to a pandas DataFrame\n",
    "df = pd.DataFrame(json_objects)\n",
    "\n",
    "# Display the DataFrame\n",
    "print(df.head())\n"
   ]
  }
 ],
 "metadata": {
  "application/vnd.databricks.v1+notebook": {
   "dashboards": [],
   "environmentMetadata": null,
   "language": "python",
   "notebookMetadata": {
    "pythonIndentUnit": 4
   },
   "notebookName": "Graphical Abstract Generator Data Pre-processing",
   "widgets": {}
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
